// @adonisjs/assembler/build/src/code_scanners/routes_scanner/main.js
import { m as debug_default, t as VirtualFileSystem } from "../../../virtual_file_system-bGeoWsK-.js";
import { n as PathsResolver, t as extractValidators } from "../../../validator_extractor-Ccio_Ndi.js";
import { cliui } from "@poppinss/cliui";
import string from "@poppinss/utils/string";
import StringBuilder from "@poppinss/utils/string_builder";
import { parseImports } from "parse-imports";
var RoutesScanner = class {
	#filter;
	#appRoot;
	#controllerRoutes = {};
	#scannedRoutes = [];
	#computeResponseTypes;
	#computeRequestTypes;
	#extractValidators;
	ui = cliui();
	pathsResolver;
	rules = {
		request: {},
		response: {}
	};
	constructor(appRoot, rulesCollection) {
		this.#appRoot = appRoot;
		this.pathsResolver = new PathsResolver(appRoot);
		rulesCollection.forEach((rules) => {
			Object.assign(this.rules.request, rules.request);
			Object.assign(this.rules.response, rules.response);
		});
	}
	#shouldSkipRoute(route) {
		if (this.#filter) return !this.#filter(route);
		return false;
	}
	#prepareRequestTypes(route) {
		if (!route.validators) return;
		return {
			type: route.validators.reduce((result, validator) => {
				const validatorExport = validator.import.type === "default" ? ".default" : validator.import.type === "named" ? `.${validator.import.value}` : "";
				const [, ...segments] = validator.name.split(".");
				const namespace = segments.map((segment) => `['${segment}']`).join("");
				result.push(`InferInput<(typeof import('${validator.import.specifier}')${validatorExport})${namespace}>`);
				return result;
			}, []).join("|"),
			imports: [`import { InferInput } from '@vinejs/vine/types'`]
		};
	}
	async #inspectControllerSpecifier(importExpression, method) {
		const importedModule = [...await parseImports(importExpression)].find(($import) => $import.isDynamicImport && $import.moduleSpecifier.value);
		if (!importedModule || importedModule.moduleSpecifier.type !== "package" || !importedModule.moduleSpecifier.value) return null;
		const specifier = importedModule.moduleSpecifier.value;
		const name = new StringBuilder(specifier.split("/").pop()).removeSuffix("Controller").pascalCase().suffix("Controller").toString();
		return {
			name,
			method: method ?? "handle",
			path: string.toUnixSlash(this.pathsResolver.resolve(specifier, true)),
			import: {
				specifier,
				type: "default",
				value: name
			}
		};
	}
	async #setResponse(route, controller) {
		route.response = await this.#computeResponseTypes?.(route, controller, this) ?? {
			type: `ReturnType<import('${controller.import.specifier}').default['${controller.method}']>`,
			imports: []
		};
		debug_default("computed route \"%s\" response %O", route.name, route.response);
	}
	async #setRequest(route, controller, vfs) {
		route.validators = await this.#extractValidators?.(route, controller, this) ?? await extractValidators(this.#appRoot, vfs, controller);
		debug_default("computed route \"%s\" validators %O", route.name, route.validators);
		route.request = await this.#computeRequestTypes?.(route, controller, this) ?? this.#prepareRequestTypes(route);
		debug_default("computed route \"%s\" request input %O", route.name, route.request);
	}
	#processRouteWithoutController(route) {
		if (!route.name) {
			debug_default(`skipping route "%s" as it does not have a name`, route.pattern);
			return;
		}
		const scannedRoute = {
			name: route.name,
			domain: route.domain,
			methods: route.methods,
			pattern: route.pattern,
			tokens: route.tokens,
			request: this.rules.request[route.name],
			response: this.rules.response[route.name]
		};
		debug_default("scanned route without controller %O", scannedRoute);
		this.#scannedRoutes.push(scannedRoute);
	}
	async #processRouteWithController(route, vfs) {
		if (!route.handler || !route.handler.importExpression) return this.#processRouteWithoutController(route);
		const controller = await this.#inspectControllerSpecifier(route.handler.importExpression, route.handler.method);
		if (!controller) return this.#processRouteWithoutController(route);
		debug_default("processing route \"%s\" with inspected controller %O", route.name, controller);
		route.name = route.name ?? new StringBuilder(controller.name).removeSuffix("Controller").snakeCase().suffix(`.${string.snakeCase(controller.method)}`).toString();
		if (this.#shouldSkipRoute(route)) return;
		const scannedRoute = {
			name: route.name,
			domain: route.domain,
			methods: route.methods,
			pattern: route.pattern,
			tokens: route.tokens,
			request: this.rules.request[route.name],
			response: this.rules.response[route.name],
			controller
		};
		debug_default("scanned route %O", scannedRoute);
		this.#scannedRoutes.push(scannedRoute);
		if (!scannedRoute.request || !scannedRoute.response) {
			debug_default("tracking controller for rescanning %O", scannedRoute);
			this.#controllerRoutes[controller.path] ??= [];
			this.#controllerRoutes[controller.path].push(scannedRoute);
			if (!scannedRoute.response) await this.#setResponse(scannedRoute, controller);
			if (!scannedRoute.request) await this.#setRequest(scannedRoute, controller, vfs);
		}
	}
	async #processRoute(route, vfs) {
		if (route.name && this.#shouldSkipRoute(route)) {
			debug_default("route skipped route: %O, rules: %O", route, this.rules);
			return;
		}
		if (typeof route.handler === "function") {
			this.#processRouteWithoutController(route);
			return;
		}
		await this.#processRouteWithController(route, vfs);
	}
	defineResponse(callback) {
		this.#computeResponseTypes = callback;
		return this;
	}
	defineRequest(callback) {
		this.#computeRequestTypes = callback;
		return this;
	}
	extractValidators(callback) {
		this.#extractValidators = callback;
		return this;
	}
	getScannedRoutes() {
		return this.#scannedRoutes;
	}
	getControllers() {
		return Object.keys(this.#controllerRoutes);
	}
	async invalidate(controllerPath) {
		const controllerRoutes = this.#controllerRoutes[controllerPath];
		if (!controllerRoutes || !controllerRoutes.length) {
			debug_default("\"%s\" controllers is not part of scanned controllers %O", controllerPath, this.#controllerRoutes);
			return false;
		}
		for (let scannedRoute of controllerRoutes) if (scannedRoute.controller) {
			debug_default("invalidating route %O", scannedRoute);
			const vfs = new VirtualFileSystem(this.#appRoot);
			await this.#setResponse(scannedRoute, scannedRoute.controller);
			await this.#setRequest(scannedRoute, scannedRoute.controller, vfs);
		}
		return true;
	}
	filter(filterFn) {
		this.#filter = filterFn;
		return this;
	}
	async scan(routes) {
		const vfs = new VirtualFileSystem(this.#appRoot);
		for (const route of routes) await this.#processRoute(route, vfs);
		vfs.invalidate();
	}
};
export { RoutesScanner };
