// @japa/browser-client/build/index.js
import { debuglog, inspect } from "node:util";
import { join } from "node:path";
import { chromium, firefox, webkit } from "playwright";
import qs from "@poppinss/qs";
import { AssertionError } from "node:assert";
import { setTimeout } from "node:timers/promises";
import { rm } from "node:fs/promises";
import slugify from "@sindresorhus/slugify";
var BasePage = class {
	constructor(page, context) {
		this.page = page;
		this.context = context;
	}
};
var debug_default = debuglog("japa:browser-client");
const DEFAULT_LAUNCHERS = {
	chromium: (launcherOptions) => chromium.launch(launcherOptions),
	firefox: (launcherOptions) => firefox.launch(launcherOptions),
	webkit: (launcherOptions) => webkit.launch(launcherOptions)
};
const DEFAULT_ASSERTIONS_CONFIG = {
	timeout: 5e3,
	pollIntervals: [
		100,
		250,
		500,
		1e3
	]
};
function normalizeConfig(cliArgs, config) {
	const tracingEvent = cliArgs?.trace;
	if (tracingEvent && !["onError", "onTest"].includes(tracingEvent)) throw new Error(`Invalid tracing event "${tracingEvent}". Use --trace="onTest" or --trace="onError"`);
	if (tracingEvent) config.tracing = Object.assign(config.tracing || {
		outputDirectory: join(process.cwd(), "./"),
		cleanOutputDirectory: true
	}, {
		enabled: true,
		event: tracingEvent
	});
	return {
		...config,
		launcher: config.launcher || (async (launcherOptions) => {
			const browser = cliArgs?.browser || "chromium";
			const launcher = DEFAULT_LAUNCHERS[browser];
			if (!launcher) throw new Error(`Invalid browser "${browser}". Allowed values are ${Object.keys(DEFAULT_LAUNCHERS).join(", ")}`);
			return launcher(launcherOptions);
		})
	};
}
function decoratePage(page, context, decorators, config) {
	decorators.forEach((decorator) => {
		if (decorator.page) decorator.page(page, context, config);
	});
	page.on("response", (response) => {
		decorators.forEach((decorator) => {
			if (decorator.response) decorator.response(response);
		});
	});
	return page;
}
function decorateContext(context, decorators, config) {
	decorators.forEach((decorator) => {
		if (decorator.context) decorator.context(context);
	});
	context.on("page", (page) => decoratePage(page, context, decorators, config));
	return context;
}
function decorateBrowser(browser, decorators, config = {}) {
	const originalNewContext = browser.newContext.bind(browser);
	const originalNewPage = browser.newPage.bind(browser);
	browser.newContext = async function(options) {
		return decorateContext(await originalNewContext(options), decorators, config);
	};
	browser.newPage = async function(...args) {
		const page = await originalNewPage(...args);
		return decoratePage(page, page.context(), decorators, config);
	};
	return browser;
}
function getLauncherOptions(cliArgs) {
	const options = {
		headless: !cliArgs?.headed,
		slowMo: cliArgs?.slow === true ? 100 : Number(cliArgs?.slow) || void 0,
		devtools: !!cliArgs?.devtools
	};
	debug_default("using launcher options %O", options);
	return options;
}
const addUseMethod = { page(page) {
	page.use = function(PageOrInteraction) {
		return new PageOrInteraction(this, this.context());
	};
} };
const addVisitMethod = { context(context) {
	context.visit = async function(UrlOrPage, callbackOrOptions) {
		const page = await context.newPage();
		if (typeof UrlOrPage === "string") {
			await page.goto(UrlOrPage, callbackOrOptions);
			return page;
		}
		const pageInstance = new UrlOrPage(page, context);
		await page.goto(pageInstance.url, pageInstance.visitOptions);
		if (typeof callbackOrOptions === "function") {
			await callbackOrOptions(pageInstance);
			return;
		}
		return pageInstance;
	};
} };
const addPauseMethods = { page(page) {
	page.pauseIf = async function(condition) {
		if (condition) await this.pause();
	};
	page.pauseUnless = async function(condition) {
		if (!condition) await this.pause();
	};
} };
const isSubsetOf = (superset, subset) => {
	if (typeof superset !== "object" || superset === null || typeof subset !== "object" || subset === null) return false;
	return Object.keys(subset).every((key) => {
		if (!superset.propertyIsEnumerable(key)) return false;
		const subsetItem = subset[key];
		const supersetItem = superset[key];
		if (typeof subsetItem === "object" && subsetItem !== null) return isSubsetOf(supersetItem, subsetItem);
		if (supersetItem !== subsetItem) return false;
		return true;
	});
};
async function retryTest(options, callback) {
	const { pollIntervals, timeout } = options;
	const remainingIntervals = pollIntervals.slice(0, -1);
	const repeatedLastInterval = pollIntervals[pollIntervals.length - 1];
	let hasStopped = false;
	let lastError;
	let lastAttemptCallback;
	const attempt = async () => {
		try {
			lastAttemptCallback = callback();
			await lastAttemptCallback;
			hasStopped = true;
			return {};
		} catch (error$1) {
			if (hasStopped) return {};
			lastError = error$1;
			await setTimeout(remainingIntervals.shift() ?? repeatedLastInterval);
			if (hasStopped) return {};
			return attempt();
		}
	};
	const failAfterTimeout = () => setTimeout(timeout).then(async () => {
		if (hasStopped) return {};
		hasStopped = true;
		if (!lastError) return { error: /* @__PURE__ */ new Error("retryTest: Callback ran out of time") };
		else {
			await lastAttemptCallback.catch((e) => lastError = e);
			return { error: lastError };
		}
	});
	const { error } = await Promise.race([attempt(), failAfterTimeout()]);
	if (error) {
		error.message += `, timed out after ${timeout}ms`;
		throw error;
	}
}
function getLocator(selector, page) {
	return typeof selector === "string" ? page.locator(selector) : selector;
}
const addAssertions = { page(page, _context, config) {
	const retrySettings = {
		...DEFAULT_ASSERTIONS_CONFIG,
		...config?.assertions
	};
	page.assertExists = function assertExists(selector) {
		return retryTest(retrySettings, async () => {
			if (await getLocator(selector, this).count() <= 0) throw new AssertionError({
				message: `expected ${inspect(selector)} element to exist`,
				stackStartFn: assertExists
			});
		});
	};
	page.assertNotExists = function assertNotExists(selector) {
		return retryTest(retrySettings, async () => {
			if (await getLocator(selector, this).count() !== 0) throw new AssertionError({
				message: `expected ${inspect(selector)} element to not exist`,
				stackStartFn: assertNotExists
			});
		});
	};
	page.assertElementsCount = function assertElementsCount(selector, expectedCount) {
		return retryTest(retrySettings, async () => {
			const matchingCount = await getLocator(selector, this).count();
			if (matchingCount !== expectedCount) throw new AssertionError({
				message: `expected ${inspect(selector)} to have '${expectedCount}' elements`,
				stackStartFn: assertElementsCount,
				actual: matchingCount,
				expected: expectedCount
			});
		});
	};
	page.assertVisible = function assertVisible(selector) {
		return retryTest(retrySettings, async () => {
			if (!await getLocator(selector, this).isVisible()) throw new AssertionError({
				message: `expected ${inspect(selector)} element to be visible`,
				stackStartFn: assertVisible
			});
		});
	};
	page.assertNotVisible = function assertNotVisible(selector) {
		return retryTest(retrySettings, async () => {
			if (await getLocator(selector, this).isVisible()) throw new AssertionError({
				message: `expected ${inspect(selector)} element to be not visible`,
				stackStartFn: assertNotVisible
			});
		});
	};
	page.assertTitle = function assertTitle(expectedTitle) {
		return retryTest(retrySettings, async () => {
			const title = await this.title();
			if (title !== expectedTitle) throw new AssertionError({
				message: `expected page title '${title}' to equal '${expectedTitle}'`,
				stackStartFn: assertTitle,
				actual: title,
				expected: expectedTitle
			});
		});
	};
	page.assertTitleContains = function assertTitleContains(expectedSubstring) {
		return retryTest(retrySettings, async () => {
			const pageTitle = await this.title();
			if (!pageTitle.includes(expectedSubstring)) throw new AssertionError({
				message: `expected page title '${pageTitle}' to include '${expectedSubstring}'`,
				stackStartFn: assertTitleContains
			});
		});
	};
	page.assertUrl = function assertUrl(expectedUrl) {
		return retryTest(retrySettings, async () => {
			const url = this.url();
			if (url !== expectedUrl) throw new AssertionError({
				message: `expected page URL '${url}' to equal '${expectedUrl}'`,
				stackStartFn: assertUrl,
				actual: url,
				expected: expectedUrl
			});
		});
	};
	page.assertUrlContains = function assertUrlContains(expectedSubstring) {
		return retryTest(retrySettings, async () => {
			const pageUrl = this.url();
			if (!pageUrl.includes(expectedSubstring)) throw new AssertionError({
				message: `expected page URL '${pageUrl}' to include '${expectedSubstring}'`,
				stackStartFn: assertUrlContains
			});
		});
	};
	page.assertUrlMatches = function assertUrlMatches(regex) {
		return retryTest(retrySettings, async () => {
			const url = this.url();
			if (!regex.test(url)) throw new AssertionError({
				message: `expected page URL '${url}' to match '${regex}'`,
				stackStartFn: assertUrlMatches,
				actual: url,
				expected: regex
			});
		});
	};
	page.assertPath = function assertPath(expectedPathName) {
		return retryTest(retrySettings, async () => {
			const { pathname } = new URL(this.url());
			if (pathname !== expectedPathName) throw new AssertionError({
				message: `expected page pathname '${pathname}' to equal '${expectedPathName}'`,
				stackStartFn: assertPath,
				actual: pathname,
				expected: expectedPathName
			});
		});
	};
	page.assertPathContains = function assertPathContains(expectedSubstring) {
		return retryTest(retrySettings, async () => {
			const { pathname } = new URL(this.url());
			if (!pathname.includes(expectedSubstring)) throw new AssertionError({
				message: `expected page pathname '${pathname}' to include '${expectedSubstring}'`,
				stackStartFn: assertPathContains
			});
		});
	};
	page.assertPathMatches = function assertPathMatches(regex) {
		return retryTest(retrySettings, async () => {
			const { pathname } = new URL(this.url());
			if (!regex.test(pathname)) throw new AssertionError({
				message: `expected page pathname '${pathname}' to match '${regex}'`,
				stackStartFn: assertPathMatches
			});
		});
	};
	page.assertQueryString = function assertQueryString(expectedSubset) {
		return retryTest(retrySettings, async () => {
			const pageURL = new URL(this.url());
			const queryString = qs.parse(pageURL.search, { ignoreQueryPrefix: true });
			if (!isSubsetOf(queryString, expectedSubset)) throw new AssertionError({
				message: `expected '${inspect(queryString)}' to contain '${inspect(expectedSubset)}'`,
				stackStartFn: assertQueryString,
				actual: queryString,
				expected: expectedSubset
			});
		});
	};
	page.assertCookie = function assertCookie(cookieName, value) {
		return retryTest(retrySettings, async () => {
			const matchingCookie = (await this.context().cookies()).find(({ name }) => name === cookieName);
			if (!matchingCookie) throw new AssertionError({
				message: `expected '${cookieName}' cookie to exist`,
				stackStartFn: assertCookie
			});
			if (value && matchingCookie.value !== value) throw new AssertionError({
				message: `expected '${cookieName}' cookie value to equal '${value}'`,
				stackStartFn: assertCookie,
				actual: matchingCookie.value,
				expected: value
			});
		});
	};
	page.assertCookieMissing = function assertCookieMissing(cookieName) {
		return retryTest(retrySettings, async () => {
			if ((await this.context().cookies()).find(({ name }) => name === cookieName)) throw new AssertionError({
				message: `expected '${cookieName}' cookie to not exist`,
				stackStartFn: assertCookieMissing
			});
		});
	};
	page.assertText = function assertText(selector, expectedValue) {
		return retryTest(retrySettings, async () => {
			const actual = await getLocator(selector, this).innerText({ timeout: 2e3 });
			if (actual !== expectedValue) throw new AssertionError({
				message: `expected ${inspect(selector)} inner text to equal '${expectedValue}'`,
				stackStartFn: assertText,
				actual,
				expected: expectedValue
			});
		});
	};
	page.assertElementsText = function assertElementsText(selector, expectedValues) {
		return retryTest(retrySettings, async () => {
			const innertTexts = await getLocator(selector, this).allInnerTexts();
			innertTexts.forEach((text, index) => {
				if (text !== expectedValues[index]) throw new AssertionError({
					message: `expected ${inspect(selector)} value to deeply equal ${inspect(expectedValues)} in same order`,
					stackStartFn: assertElementsText,
					actual: innertTexts,
					expected: expectedValues
				});
			});
		});
	};
	page.assertTextContains = function assertTextContains(selector, expectedSubstring) {
		return retryTest(retrySettings, async () => {
			if (!(await getLocator(selector, this).innerText({ timeout: 2e3 })).includes(expectedSubstring)) throw new AssertionError({
				message: `expected ${inspect(selector)} inner text to include '${expectedSubstring}'`,
				stackStartFn: assertTextContains
			});
		});
	};
	page.assertChecked = function assertChecked(selector) {
		return retryTest(retrySettings, async () => {
			let isChecked;
			try {
				isChecked = await getLocator(selector, this).isChecked({ timeout: 2e3 });
			} catch (error) {
				if (error.message.includes("Not a checkbox")) throw new AssertionError({
					message: `expected ${inspect(selector)} to be a checkbox`,
					stackStartFn: assertChecked
				});
				throw error;
			}
			if (isChecked === false) throw new AssertionError({
				message: `expected ${inspect(selector)} checkbox to be checked`,
				stackStartFn: assertChecked
			});
		});
	};
	page.assertNotChecked = function assertNotChecked(selector) {
		return retryTest(retrySettings, async () => {
			let isChecked;
			try {
				isChecked = await getLocator(selector, this).isChecked({ timeout: 2e3 });
			} catch (error) {
				if (error.message.includes("Not a checkbox")) throw new AssertionError({
					message: `expected ${inspect(selector)} to be a checkbox`,
					stackStartFn: assertNotChecked
				});
				throw error;
			}
			if (isChecked === true) throw new AssertionError({
				message: `expected ${inspect(selector)} checkbox to be not checked`,
				stackStartFn: assertNotChecked
			});
		});
	};
	page.assertDisabled = function assertDisabled(selector) {
		return retryTest(retrySettings, async () => {
			if (!await getLocator(selector, this).isDisabled({ timeout: 2e3 })) throw new AssertionError({
				message: `expected ${inspect(selector)} element to be disabled`,
				stackStartFn: assertDisabled
			});
		});
	};
	page.assertNotDisabled = function assertNotDisabled(selector) {
		return retryTest(retrySettings, async () => {
			if (await getLocator(selector, this).isDisabled({ timeout: 2e3 })) throw new AssertionError({
				message: `expected ${inspect(selector)} element to be not disabled`,
				stackStartFn: assertNotDisabled
			});
		});
	};
	page.assertInputValue = function assertInputValue(selector, expectedValue) {
		return retryTest(retrySettings, async () => {
			let inputValue;
			try {
				inputValue = await getLocator(selector, this).inputValue({ timeout: 2e3 });
			} catch (error) {
				if (error.message.includes("Node is not")) throw new AssertionError({
					message: `expected ${inspect(selector)} element to be an input, select or a textarea`,
					stackStartFn: assertInputValue
				});
				throw error;
			}
			if (inputValue !== expectedValue) throw new AssertionError({
				message: `expected ${inspect(selector)} value to equal '${expectedValue}'`,
				stackStartFn: assertInputValue,
				actual: inputValue,
				expected: expectedValue
			});
		});
	};
	page.assertSelectedOptions = function assertSelectedOptions(selector, expectedValues) {
		return retryTest(retrySettings, async () => {
			const element = await this.$eval(selector, (node) => {
				/* c8 ignore start */
				if (node.nodeName === "SELECT") {
					const options = [...node.options];
					return {
						multiple: node.multiple,
						selected: options.filter((option) => option.selected).map((option) => option.value)
					};
				}
				/* c8 ignore end */
			});
			if (!element) throw new AssertionError({
				message: `expected ${inspect(selector)} element to be a select box`,
				stackStartFn: assertSelectedOptions
			});
			if (element.multiple) element.selected.forEach((elem) => {
				if (!expectedValues.includes(elem)) throw new AssertionError({
					message: `expected ${inspect(selector)} value to deeply equal ${inspect(expectedValues)}`,
					stackStartFn: assertSelectedOptions,
					actual: element.selected,
					expected: expectedValues
				});
			});
			else if (element.selected[0] !== expectedValues[0]) throw new AssertionError({
				message: `expected ${inspect(selector)} value to equal ${inspect(expectedValues)}`,
				stackStartFn: assertSelectedOptions,
				actual: element.selected[0],
				expected: expectedValues[0]
			});
		});
	};
} };
var DecoratorsCollection = class {
	#list = [
		addAssertions,
		addPauseMethods,
		addUseMethod,
		addVisitMethod
	];
	register(decorator) {
		this.#list.push(decorator);
		return this;
	}
	toList() {
		return this.#list;
	}
};
const decoratorsCollection = new DecoratorsCollection();
async function traceActionsHook(tracingConfig, test) {
	const suiteName = test.options.meta.suite.name;
	await test.context.browserContext.tracing.start({
		title: test.title,
		screenshots: true,
		snapshots: true,
		sources: true
	});
	if (tracingConfig.event === "onError") return async (hasError) => {
		if (hasError) await test.context.browserContext.tracing.stop({ path: join(tracingConfig.outputDirectory, suiteName, `${slugify(test.title)}.zip`) });
		else await test.context.browserContext.tracing.stop();
	};
	return async () => {
		await test.context.browserContext.tracing.stop({ path: join(tracingConfig.outputDirectory, suiteName, `${slugify(test.title)}.zip`) });
	};
}
async function cleanTracesHook(suite, outputDirectory) {
	const suiteDirectory = join(outputDirectory, suite.name);
	debug_default("removing traces output from %s location", suiteDirectory);
	await rm(suiteDirectory, {
		recursive: true,
		force: true,
		maxRetries: 4
	});
}
var BrowserProxy = class {
	constructor(suite) {
		return new Proxy(this, { get(_, property) {
			throw new Error(`Cannot access "browser.${String(property)}". The browser is not configured to run for "${suite}" suite`);
		} });
	}
};
var BrowserContextProxy = class {
	constructor(suite) {
		return new Proxy(this, { get(_, property) {
			throw new Error(`Cannot access "browserContext.${String(property)}". The browser is not configured to run for "${suite}" suite`);
		} });
	}
};
async function createContextHook(browser, config, test) {
	const host = process.env.HOST;
	const port = process.env.PORT;
	const context = test.context;
	debug_default("creating browser context for test \"%s\"", context.test.title);
	context.browser = browser;
	context.browserContext = await browser.newContext({
		baseURL: host && port ? `http://${host}:${port}` : void 0,
		...config.contextOptions
	});
	context.visit = context.browserContext.visit.bind(context.browserContext);
	context.record = async function(urlOrCallback, options) {
		await this.browserContext._enableRecorder.bind(this.browserContext)({
			mode: "recording",
			handleSIGINT: false
		});
		(typeof urlOrCallback === "function" ? await urlOrCallback() : await this.visit(urlOrCallback, options)).on("close", async () => {
			await this.browserContext.close();
			await this.browser.close();
		});
		return new Promise((resolve) => {
			this.browser.on("disconnected", () => resolve());
		});
	}.bind(context);
	return () => {
		debug_default("closing browser context for test \"%s\"", context.test.title);
		return context.browserContext.close();
	};
}
function createFakeContextHook(test) {
	const suiteName = test.options.meta.suite.name;
	test.context.browser = new BrowserProxy(suiteName);
	test.context.browserContext = new BrowserContextProxy(suiteName);
	test.context.visit = async function() {
		throw new Error(`Cannot access call "visit". The browser is not configured to run for "${suiteName}" suite`);
	};
}
function browserClient(config) {
	const clientPlugin = function(japa) {
		const normalizedConfig = normalizeConfig(japa.cliArgs, config);
		const launcherOptions = getLauncherOptions(japa.cliArgs);
		japa.runner.onSuite((suite) => {
			let browser;
			const shouldInitiateBrowser = !config.runInSuites || config.runInSuites.includes(suite.name);
			const isTracingEnabled = normalizedConfig.tracing && normalizedConfig.tracing.enabled;
			const shouldCleanOutputDirectory = isTracingEnabled && normalizedConfig.tracing.cleanOutputDirectory && normalizedConfig.tracing.outputDirectory;
			if (shouldInitiateBrowser) {
				if (shouldCleanOutputDirectory) suite.setup(() => cleanTracesHook(suite, normalizedConfig.tracing.outputDirectory));
				suite.setup(async () => {
					debug_default("initiating browser for suite \"%s\"", suite.name);
					browser = decorateBrowser(await normalizedConfig.launcher(launcherOptions), decoratorsCollection.toList(), config);
					return () => {
						debug_default("closing browser for suite \"%s\"", suite.name);
						return browser.close();
					};
				});
			}
			suite.onGroup((group) => {
				if (shouldInitiateBrowser) {
					group.each.setup((test) => createContextHook(browser, config, test));
					if (isTracingEnabled) group.each.setup((test) => traceActionsHook(normalizedConfig.tracing, test));
				} else group.each.setup((test) => createFakeContextHook(test));
			});
			suite.onTest((test) => {
				if (shouldInitiateBrowser) {
					test.setup((self) => createContextHook(browser, config, self));
					if (isTracingEnabled) test.setup((self) => traceActionsHook(normalizedConfig.tracing, self));
				} else test.setup((self) => createFakeContextHook(self));
			});
		});
	};
	return clientPlugin;
}
var BaseInteraction = class {
	#queue = /* @__PURE__ */ new Set();
	constructor(page, context) {
		this.page = page;
		this.context = context;
	}
	defer(action) {
		this.#queue.add(action);
		return this;
	}
	async exec() {
		for (let action of this.#queue) await action();
	}
};
export { BaseInteraction, BasePage, browserClient, decorateBrowser, decoratorsCollection };
