// @vinejs/vine/build/factories/main.js
import { i as SimpleMessagesProvider, s as helpers, t as SimpleErrorReporter } from "../simple_error_reporter-DjITOpes.js";
import { AssertionError, deepEqual } from "node:assert";
var FieldFactory = class {
	create(fieldName, value, messagesProvider, errorReporter) {
		const reporter = errorReporter || new SimpleErrorReporter();
		const provider = messagesProvider || new SimpleMessagesProvider({}, {});
		return {
			value,
			isArrayMember: false,
			parent: { [fieldName]: value },
			data: { [fieldName]: value },
			name: fieldName,
			wildCardPath: fieldName,
			getFieldPath() {
				return fieldName;
			},
			isDefined: helpers.exists(value),
			isValid: true,
			isValidDataType: false,
			meta: {},
			mutate(newValue) {
				this.value = newValue;
				this.isDefined = helpers.exists(newValue);
				return this;
			},
			report(message, rule, context, args) {
				this.isValid = false;
				reporter.report(provider.getMessage(message, rule, context, args), rule, context, args);
			}
		};
	}
};
var ValidationResult = class {
	#outputValue;
	#reporter;
	constructor(outputValue, reporter) {
		this.#outputValue = outputValue;
		this.#reporter = reporter;
	}
	#assertionError(options) {
		const assertion = new AssertionError(options);
		Object.defineProperty(assertion, "showDiff", { value: true });
		return assertion;
	}
	getOutput() {
		return this.#outputValue;
	}
	getErrors() {
		return this.#reporter.errors;
	}
	assertSucceeded() {
		if (this.#reporter.hasErrors) {
			const errorsCount = this.#reporter.errors.length;
			throw this.#assertionError({
				message: `Expected validation to pass. Instead failed with "${errorsCount} error(s)"`,
				operator: "strictEqual",
				stackStartFn: this.assertSucceeded
			});
		}
	}
	assertOutput(expectedOutput) {
		deepEqual(this.#outputValue, expectedOutput);
	}
	assertFailed() {
		if (!this.#reporter.hasErrors) throw this.#assertionError({
			message: `Expected validation to report one or more errors`,
			operator: "strictEqual",
			stackStartFn: this.assertFailed
		});
	}
	assertErrorsCount(count) {
		const errorsCount = this.#reporter.errors.length;
		if (errorsCount !== count) throw this.#assertionError({
			message: `Expected validation to report "${count}" errors. Received "${errorsCount}"`,
			expected: count,
			actual: errorsCount,
			operator: "strictEqual",
			stackStartFn: this.assertErrorsCount,
			showDiff: true
		});
	}
	assertError(message) {
		const messages = this.#reporter.errors.map((e) => e.message);
		if (!messages.includes(message)) throw this.#assertionError({
			message: `Expected validation errors to include "${message}" message`,
			expected: [message],
			actual: messages,
			operator: "includes",
			stackStartFn: this.assertError,
			showDiff: true
		});
	}
};
var ValidatorFactory = class ValidatorFactory {
	#field;
	#bail;
	#dataTypeValidation;
	constructor(field, bail, dataTypeValidation) {
		this.#field = field;
		this.#bail = bail;
		this.#dataTypeValidation = dataTypeValidation;
	}
	#getReporter() {
		return new SimpleErrorReporter();
	}
	withContext(field) {
		return new ValidatorFactory(field, this.#bail, this.#dataTypeValidation);
	}
	withDataTypeValidator(validation) {
		return new ValidatorFactory(this.#field, this.#bail, validation);
	}
	bail(state) {
		return new ValidatorFactory(this.#field, state, this.#dataTypeValidation);
	}
	execute(validation, value) {
		const errorReporter = this.#getReporter();
		const bail = this.#bail === false ? false : true;
		const field = {
			...new FieldFactory().create("dummy", value, void 0, errorReporter),
			...this.#field
		};
		let continuationCheck = field.isDefined;
		if (this.#dataTypeValidation) {
			field.isValidDataType = this.#dataTypeValidation.rule.validator(field.value, this.#dataTypeValidation.options, field);
			continuationCheck = field.isValidDataType;
		}
		const validations = Array.isArray(validation) ? validation : [validation];
		for (let one of validations) {
			if (one.rule.isAsync) throw new Error(`Cannot execute async rule "${one.rule.validator.name}". Use "validator.executeAsync" instead`);
			if ((continuationCheck || one.rule.implicit) && (field.isValid || !bail)) one.rule.validator(field.value, one.options, field);
		}
		return new ValidationResult(field.value, errorReporter);
	}
	async executeAsync(validation, value) {
		const errorReporter = this.#getReporter();
		const bail = this.#bail === false ? false : true;
		const field = {
			...new FieldFactory().create("dummy", value, void 0, errorReporter),
			...this.#field
		};
		let continuationCheck = field.isDefined;
		if (this.#dataTypeValidation) {
			field.isValidDataType = this.#dataTypeValidation.rule.validator(field.value, this.#dataTypeValidation.options, field);
			continuationCheck = field.isValidDataType;
		}
		const validations = Array.isArray(validation) ? validation : [validation];
		for (let one of validations) if ((continuationCheck || one.rule.implicit) && (field.isValid || !bail)) await one.rule.validator(field.value, one.options, field);
		return new ValidationResult(field.value, errorReporter);
	}
};
const validator = new ValidatorFactory();
const fieldContext = new FieldFactory();
export { FieldFactory, ValidatorFactory, fieldContext, validator };
