// @adonisjs/vite/build/index.js
import {
  Vite
} from "./chunk-ZCIGOANY.js";
import "./chunk-ZCQTQOMI.js";

// stubs/main.ts
var stubsRoot = import.meta.dirname;

// configure.ts
async function configure(command) {
  const codemods = await command.createCodemods();
  let shouldInstallPackages = command.parsedFlags.install;
  await codemods.makeUsingStub(stubsRoot, "config/vite.stub", {});
  await codemods.makeUsingStub(stubsRoot, "vite.config.stub", {});
  await codemods.makeUsingStub(stubsRoot, "js_entrypoint.stub", {});
  await codemods.updateRcFile((rcFile) => {
    rcFile.addProvider("@adonisjs/vite/vite_provider");
    rcFile.addMetaFile("public/**", false);
    rcFile.addAssemblerHook("buildStarting", "@adonisjs/vite/build_hook");
  });
  await codemods.registerMiddleware("server", [
    { path: "@adonisjs/vite/vite_middleware", position: "after" }
  ]);
  if (shouldInstallPackages === void 0) {
    shouldInstallPackages = await command.prompt.confirm('Do you want to install "vite"?');
  }
  if (shouldInstallPackages) {
    await codemods.installPackages([{ name: "vite", isDevDependency: true }]);
  } else {
    await codemods.listPackagesToInstall([{ name: "vite", isDevDependency: true }]);
  }
}

// src/define_config.ts
import { join } from "path";
function defineConfig(config) {
  return {
    assetsUrl: "/assets",
    buildDirectory: "public/assets",
    manifestFile: config.buildDirectory ? join(config.buildDirectory, ".vite/manifest.json") : "public/assets/.vite/manifest.json",
    ...config
  };
}
export {
  Vite,
  configure,
  defineConfig,
  stubsRoot
};
