// @poppinss/ts-exec/build/src/loader.js
// src/loader.ts
import { extname } from "path";
import { fileURLToPath as fileURLToPath2 } from "url";

// src/debug.ts
import { debuglog } from "util";
var debug_default = debuglog("ts-exec");

// src/get_config.ts
import { fileURLToPath } from "url";
import { getTsconfig } from "get-tsconfig";
var DEFAULT_CONFIG = {
  sourceMaps: true,
  module: {
    type: "es6"
  },
  jsc: {
    target: "esnext",
    externalHelpers: false,
    parser: {
      syntax: "typescript",
      tsx: true,
      decorators: true,
      dynamicImport: true
    },
    transform: {
      decoratorMetadata: true,
      legacyDecorator: true,
      verbatimModuleSyntax: true,
      useDefineForClassFields: true,
      react: {
        runtime: "automatic"
      }
    },
    experimental: {
      keepImportAttributes: true
    },
    keepClassNames: true
  }
};
function tsConfigToSwcConfig(tsConfig2) {
  const tsTarget = tsConfig2.compilerOptions?.target?.toLowerCase();
  return {
    sourceMaps: tsConfig2.compilerOptions?.sourceMap ?? true,
    module: {
      /**
       * Since, we will run the SWC transformer via Nodejs loader hooks, the
       * entry point will always be ES6
       */
      type: "es6"
    },
    jsc: {
      target: tsTarget ? tsTarget === "es6" ? "esnext" : tsTarget : "esnext",
      baseUrl: tsConfig2.compilerOptions?.baseUrl,
      externalHelpers: false,
      parser: {
        syntax: "typescript",
        tsx: !!tsConfig2.compilerOptions?.jsx,
        decorators: true,
        dynamicImport: true
      },
      transform: {
        ...tsConfig2.compilerOptions?.emitDecoratorMetadata ? {
          decoratorMetadata: true
        } : {},
        ...tsConfig2.compilerOptions?.experimentalDecorators ? {
          legacyDecorator: true
        } : {},
        ...tsConfig2.compilerOptions?.verbatimModuleSyntax ? {
          verbatimModuleSyntax: true
        } : {},
        ...tsConfig2.compilerOptions?.useDefineForClassFields ? {
          useDefineForClassFields: true
        } : {},
        react: {
          runtime: "automatic",
          ...tsConfig2.compilerOptions?.jsxImportSource ? {
            importSource: tsConfig2.compilerOptions?.jsxImportSource
          } : {}
        }
      },
      experimental: {
        keepImportAttributes: true
      },
      keepClassNames: true
    }
  };
}
function getConfig(searchPath) {
  const tsConfig2 = getTsconfig(
    typeof searchPath === "string" ? searchPath : fileURLToPath(searchPath)
  );
  if (!tsConfig2) {
    debug_default("Unable to locate tsconfig file");
    return {
      tsConfig: null,
      tsConfigPath: null,
      swcConfig: {
        ...DEFAULT_CONFIG
      }
    };
  }
  debug_default("tsConfig %O", tsConfig2.config);
  const swcConfig2 = tsConfigToSwcConfig(tsConfig2.config);
  debug_default("swcConfig %O", swcConfig2);
  return { tsConfig: tsConfig2.config, tsConfigPath: tsConfig2.path, swcConfig: swcConfig2 };
}

// src/transform.ts
import { transform as swcTransform } from "@swc/core";
function transformSync(source, options) {
  const input = `${source ?? ""}`;
  const { format, ...rest } = options;
  rest.module.type = format === "commonjs-typescript" ? "commonjs" : "es6";
  return swcTransform(input, rest);
}

// src/loader.ts
var swcConfig;
var tsConfig;
function isTypeScriptFile(filePath) {
  return !filePath.endsWith(".d.ts") && (filePath.endsWith(".ts") || filePath.endsWith(".tsx") || filePath.endsWith(".mts") || filePath.endsWith(".cts"));
}
function isLocalPath(specifier) {
  return specifier.startsWith("./") || specifier.startsWith("../") || specifier.startsWith("/") || specifier.startsWith("file:");
}
function wrapAndReThrowSwcError(error) {
  const errorHints = `${error.filename}:${error.startLine}
${error.snippet}
`;
  switch (error.code) {
    case "UnsupportedSyntax": {
      const unsupportedSyntaxError = new Error(error.message);
      unsupportedSyntaxError.name = "UnsupportedSyntaxError";
      unsupportedSyntaxError.stack = `${errorHints}${unsupportedSyntaxError.stack}`;
      throw unsupportedSyntaxError;
    }
    case "InvalidSyntax": {
      const syntaxError = new SyntaxError(error.message);
      syntaxError.stack = `${errorHints}${syntaxError.stack}`;
      throw syntaxError;
    }
    default:
      throw new Error(error.message);
  }
}
async function initialize() {
  const searchPath = process.env.TS_EXEC_PWD ?? process.cwd();
  const config = getConfig(searchPath);
  swcConfig = config.swcConfig;
  tsConfig = config.tsConfig;
}
var resolve = async (specifier, context, nextResolve) => {
  debug_default("resolving specifier %s", specifier);
  const [pathname] = specifier.split("?");
  if (!tsConfig?.compilerOptions?.rewriteRelativeImportExtensions && context.parentURL && isTypeScriptFile(pathname)) {
    throw new Error(
      `Cannot import "${specifier}" using "${extname(pathname)}" extension. Enable "compilerOptions.rewriteRelativeImportExtensions" to import TypeScript files`
    );
  }
  try {
    return await nextResolve(specifier, context);
  } catch (error) {
    if (error.code === "ERR_MODULE_NOT_FOUND" && error.url && isLocalPath(error.url)) {
      let [urlPathname, ...urlQs] = error.url.split("?");
      const fileExtension = extname(urlPathname);
      const testExtensions = [];
      if ([".js", ".mjs", ".cjs", ".jsx"].includes(fileExtension)) {
        testExtensions.push(fileExtension.replace("js", "ts"));
      }
      if (fileExtension === ".js") {
        testExtensions.push(".tsx");
      }
      for (const ext of testExtensions) {
        const tryPath = urlPathname.slice(0, -fileExtension.length) + ext;
        try {
          const url = urlQs.length ? `${tryPath}?${urlQs.join("?")}` : tryPath;
          debug_default("retrying to resolve file with specifier %s", url);
          return await nextResolve(url, context);
        } catch {
        }
      }
    }
    throw error;
  }
};
var load = async function load2(url, context, nextLoad) {
  const { format } = context;
  debug_default("load file %s, %O", url, context);
  if (format && (format === "module-typescript" || format === "commonjs-typescript") || url.endsWith(".tsx") || url.endsWith(".jsx")) {
    try {
      const { source } = await nextLoad(url, {
        ...context,
        format: "module"
      });
      const { code, map } = await transformSync(source.toString(), {
        format: format ?? "module-tsx",
        filename: fileURLToPath2(url),
        ...swcConfig
      });
      let output = code;
      if (map) {
        const base64SourceMap = Buffer.from(map).toString("base64");
        output = `${code}

      }
      return {
        format: format ? format.replace("-typescript", "") : "module",
        source: `${output}

      };
    } catch (error) {
      if (error.code !== void 0) {
        wrapAndReThrowSwcError(error);
      }
      throw error;
    }
  }
  return nextLoad(url, context);
};
export {
  initialize,
  load,
  resolve
};
