// @vinejs/vine/build/index.js
import { a as BOOLEAN_NEGATIVES, c as __exportAll, i as SimpleMessagesProvider, n as main_exports, o as BOOLEAN_POSITIVES, r as ValidationError, s as helpers, t as SimpleErrorReporter } from "./simple_error_reporter-DjITOpes.js";
import { fields, globalTransforms, messages } from "./src/defaults.js";
import dayjs from "dayjs";
import Macroable from "@poppinss/macroable";
import camelcase from "camelcase";
import normalizeUrl from "normalize-url";
import escapeValue from "validator/lib/escape.js";
import normalizeEmailValue from "validator/lib/normalizeEmail.js";
import { Compiler, refsBuilder } from "@vinejs/compiler";
function createRule(validator, metaData) {
	const rule = {
		validator,
		name: metaData?.name ?? validator.name,
		isAsync: metaData?.isAsync || validator.constructor.name === "AsyncFunction",
		implicit: metaData?.implicit ?? false,
		toJSONSchema: metaData?.toJSONSchema
	};
	return function(...options) {
		return {
			rule,
			options: options[0]
		};
	};
}
var symbols_exports = /* @__PURE__ */ __exportAll({
	COTYPE: () => COTYPE,
	IS_OF_TYPE: () => IS_OF_TYPE,
	ITYPE: () => ITYPE,
	OTYPE: () => OTYPE,
	PARSE: () => PARSE,
	SUBTYPE: () => SUBTYPE,
	UNIQUE_NAME: () => UNIQUE_NAME,
	VALIDATION: () => VALIDATION
});
const UNIQUE_NAME = Symbol.for("schema_name");
const IS_OF_TYPE = Symbol.for("is_of_type");
const PARSE = Symbol.for("parse");
const ITYPE = Symbol.for("opaque_input_type");
const OTYPE = Symbol.for("opaque_type");
const COTYPE = Symbol.for("camelcase_opaque_type");
const VALIDATION = Symbol.for("to_validation");
const SUBTYPE = Symbol.for("subtype");
const requiredWhen = createRule(function requiredWhen(_, checker, field) {
	const shouldBeRequired = checker(field);
	if (!field.isDefined && shouldBeRequired) field.report(messages.required, "required", field);
}, { implicit: true });
var ConditionalValidations = class {
	requiredWhen(otherField, operator, expectedValue) {
		if (typeof otherField === "function") return this.use(requiredWhen(otherField));
		let checker;
		switch (operator) {
			case "=":
				checker = (value) => value === expectedValue;
				break;
			case "!=":
				checker = (value) => value !== expectedValue;
				break;
			case "in":
				checker = (value) => expectedValue.includes(value);
				break;
			case "notIn":
				checker = (value) => !expectedValue.includes(value);
				break;
			case ">":
				checker = (value) => value > expectedValue;
				break;
			case "<":
				checker = (value) => value < expectedValue;
				break;
			case ">=":
				checker = (value) => value >= expectedValue;
				break;
			case "<=": checker = (value) => value <= expectedValue;
		}
		return this.use(requiredWhen((field) => {
			const otherFieldValue = helpers.getNestedValue(otherField, field);
			return checker(otherFieldValue);
		}));
	}
	requiredIfExists(fields) {
		const fieldsToExist = Array.isArray(fields) ? fields : [fields];
		return this.use(requiredWhen((field) => {
			return fieldsToExist.every((otherField) => {
				return helpers.exists(helpers.getNestedValue(otherField, field));
			});
		}));
	}
	requiredIfAnyExists(fields) {
		return this.use(requiredWhen((field) => {
			return fields.some((otherField) => helpers.exists(helpers.getNestedValue(otherField, field)));
		}));
	}
	requiredIfMissing(fields) {
		const fieldsToExist = Array.isArray(fields) ? fields : [fields];
		return this.use(requiredWhen((field) => {
			return fieldsToExist.every((otherField) => helpers.isMissing(helpers.getNestedValue(otherField, field)));
		}));
	}
	requiredIfAnyMissing(fields) {
		return this.use(requiredWhen((field) => {
			return fields.some((otherField) => helpers.isMissing(helpers.getNestedValue(otherField, field)));
		}));
	}
};
var NullableModifier$1 = class NullableModifier$1 {
	allowNull = true;
	get isOptional() {
		return this.#parent.isOptional;
	}
	#parent;
	constructor(parent) {
		this.#parent = parent;
	}
	clone() {
		return new NullableModifier$1(this.#parent.clone());
	}
	optional() {
		return new OptionalModifier$1(this);
	}
	transform(transformer) {
		return new TransformModifier(transformer, this);
	}
	meta(meta) {
		return new MetaModifier$1(this, meta);
	}
	toJSONSchema() {
		const schema = this.#parent.toJSONSchema?.() ?? {};
		if (schema.anyOf) {
			schema.anyOf.push({ type: "null" });
			return schema;
		}
		if (schema.enum) return { anyOf: [schema, { type: "null" }] };
		if (schema.type === void 0) {
			schema.type = "null";
			return schema;
		}
		if (typeof schema.type === "string") {
			schema.type = [schema.type, "null"];
			return schema;
		}
		if (Array.isArray(schema.type)) {
			schema.type.push("null");
			return schema;
		}
		return schema;
	}
	[PARSE](propertyName, refs, options) {
		const output = this.#parent[PARSE](propertyName, refs, options);
		output.allowNull = true;
		return output;
	}
};
var MetaModifier$1 = class MetaModifier$1 {
	get isOptional() {
		return this.#parent.isOptional;
	}
	get allowNull() {
		return this.#parent.allowNull;
	}
	#parent;
	#meta;
	constructor(parent, meta) {
		this.#parent = parent;
		this.#meta = meta;
	}
	optional() {
		return new OptionalModifier$1(this);
	}
	transform(transformer) {
		return new TransformModifier(transformer, this);
	}
	nullable() {
		return new NullableModifier$1(this);
	}
	clone() {
		return new MetaModifier$1(this.#parent.clone(), this.#meta);
	}
	toJSONSchema() {
		return {
			...this.#parent.toJSONSchema?.() ?? {},
			...this.#meta
		};
	}
	[PARSE](propertyName, refs, options) {
		const output = this.#parent[PARSE](propertyName, refs, options);
		output.allowNull = true;
		return output;
	}
};
var OptionalModifier$1 = class OptionalModifier$1 extends ConditionalValidations {
	isOptional = true;
	get allowNull() {
		return this.#parent.allowNull;
	}
	#parent;
	validations;
	constructor(parent, validations) {
		super();
		this.#parent = parent;
		this.validations = validations || [];
	}
	cloneValidations() {
		return this.validations.map((validation) => {
			return {
				options: validation.options,
				rule: validation.rule
			};
		});
	}
	compileValidations(refs) {
		return this.validations.map((validation) => {
			return {
				ruleFnId: refs.track({
					validator: validation.rule.validator,
					options: validation.options
				}),
				name: validation.rule.name,
				implicit: validation.rule.implicit,
				isAsync: validation.rule.isAsync
			};
		});
	}
	nullable() {
		return new NullableModifier$1(this);
	}
	transform(transformer) {
		return new TransformModifier(transformer, this);
	}
	use(validation) {
		this.validations.push(VALIDATION in validation ? validation[VALIDATION]() : validation);
		return this;
	}
	clone() {
		return new OptionalModifier$1(this.#parent.clone(), this.cloneValidations());
	}
	meta(meta) {
		return new MetaModifier$1(this, meta);
	}
	toJSONSchema() {
		return { ...this.#parent.toJSONSchema() };
	}
	[PARSE](propertyName, refs, options) {
		const output = this.#parent[PARSE](propertyName, refs, options);
		output.isOptional = true;
		output.validations = output.validations.concat(this.compileValidations(refs));
		return output;
	}
};
var TransformModifier = class TransformModifier {
	get isOptional() {
		return this.#parent.isOptional;
	}
	get allowNull() {
		return this.#parent.allowNull;
	}
	#parent;
	#transform;
	constructor(transform, parent) {
		this.#transform = transform;
		this.#parent = parent;
	}
	clone() {
		return new TransformModifier(this.#transform, this.#parent.clone());
	}
	optional() {
		return new OptionalModifier$1(this);
	}
	nullable() {
		return new NullableModifier$1(this);
	}
	meta(meta) {
		return new MetaModifier$1(this, meta);
	}
	toJSONSchema() {
		return this.#parent.toJSONSchema();
	}
	[PARSE](propertyName, refs, options) {
		const output = this.#parent[PARSE](propertyName, refs, options);
		output.transformFnId = refs.trackTransformer(this.#transform);
		return output;
	}
};
var BaseLiteralType = class extends Macroable {
	dataTypeValidator;
	options;
	validations;
	constructor(options, validations) {
		super();
		this.options = {
			bail: true,
			allowNull: false,
			isOptional: false,
			...options
		};
		this.validations = validations || [];
	}
	cloneValidations() {
		return this.validations.map((validation) => {
			return {
				options: validation.options,
				rule: validation.rule
			};
		});
	}
	cloneOptions() {
		return { ...this.options };
	}
	compileValidation(validation, refs) {
		return {
			ruleFnId: refs.track({
				validator: validation.rule.validator,
				options: validation.options
			}),
			name: validation.rule.name,
			implicit: validation.rule.implicit,
			isAsync: validation.rule.isAsync
		};
	}
	compileValidations(refs) {
		return this.validations.map((validation) => this.compileValidation(validation, refs));
	}
	toJSONSchema() {
		const schema = {};
		if (this.dataTypeValidator?.rule.toJSONSchema) this.dataTypeValidator.rule.toJSONSchema(schema, this.dataTypeValidator.options);
		for (const validation of this.validations) {
			if (!validation.rule.toJSONSchema) continue;
			validation.rule.toJSONSchema(schema, validation.options);
		}
		return schema;
	}
	parse(callback) {
		this.options.parse = callback;
		return this;
	}
	use(validation) {
		this.validations.push(VALIDATION in validation ? validation[VALIDATION]() : validation);
		return this;
	}
	bail(state) {
		this.options.bail = state;
		return this;
	}
	optional() {
		return new OptionalModifier$1(this);
	}
	nullable() {
		return new NullableModifier$1(this);
	}
	meta(meta) {
		return new MetaModifier$1(this, meta);
	}
	transform(transformer) {
		return new TransformModifier(transformer, this);
	}
	[PARSE](propertyName, refs, options) {
		return {
			type: "literal",
			subtype: this[SUBTYPE],
			fieldName: propertyName,
			...this.dataTypeValidator ? { dataTypeValidatorFnId: this.compileValidation(this.dataTypeValidator, refs).ruleFnId } : {},
			propertyName: options.toCamelCase ? camelcase(propertyName) : propertyName,
			bail: this.options.bail,
			allowNull: this.options.allowNull,
			isOptional: this.options.isOptional,
			parseFnId: this.options.parse ? refs.trackParser(this.options.parse) : void 0,
			validations: this.compileValidations(refs)
		};
	}
};
var VineAny = class VineAny extends BaseLiteralType {
	constructor(options, validations) {
		super(options, validations);
	}
	[SUBTYPE] = "any";
	clone() {
		return new VineAny(this.cloneOptions(), this.cloneValidations());
	}
	toJSONSchema() {
		const schema = { anyOf: [
			{ type: "string" },
			{ type: "number" },
			{ type: "boolean" },
			{ type: "array" },
			{ type: "object" }
		] };
		for (const validation of this.validations) {
			if (!validation.rule.toJSONSchema) continue;
			validation.rule.toJSONSchema(schema, validation.options);
		}
		return schema;
	}
};
const enumRule = createRule(function enumList(value, options, field) {
	const choices = typeof options.choices === "function" ? options.choices(field) : options.choices;
	if (!choices.includes(value)) field.report(messages.enum, "enum", field, { choices });
}, { toJSONSchema: (schema, options) => {
	if (typeof options.choices === "function") return;
	schema.enum = options.choices;
} });
var VineEnum = class VineEnum extends BaseLiteralType {
	static rules = { enum: enumRule };
	#values;
	[SUBTYPE] = "enum";
	getChoices() {
		return this.#values;
	}
	constructor(values, options, validations) {
		super(options, validations || [enumRule({ choices: values })]);
		this.#values = values;
	}
	clone() {
		return new VineEnum(this.#values, this.cloneOptions(), this.cloneValidations());
	}
};
const dateRule = createRule(function date(value, options, field) {
	if (!field.isDefined) return false;
	if (typeof value !== "string" && typeof value !== "number") {
		field.report(messages.date, "date", field);
		return false;
	}
	const { dateTime, formats } = helpers.asDayJS(value, options.formats);
	if (!dateTime.isValid()) {
		field.report(messages.date, "date", field);
		return false;
	}
	field.$value = dateTime;
	field.$formats = formats;
	field.mutate(globalTransforms.date ? globalTransforms.date(dateTime.toDate()) : dateTime.toDate(), field);
	return true;
});
const equalsRule$1 = createRule(function equals(_, options, field) {
	const compare = options.compare || "day";
	const format = options.format || field.$formats;
	const dateTime = field.$value;
	const expectedValue = typeof options.expectedValue === "function" ? options.expectedValue(field) : options.expectedValue;
	const { dateTime: expectedDateTime } = helpers.asDayJS(expectedValue, format);
	if (!expectedDateTime.isValid()) throw new Error(`Invalid datetime value "${expectedValue}" provided to the equals rule`);
	if (!dateTime.isSame(expectedDateTime, compare)) field.report(messages["date.equals"], "date.equals", field, {
		expectedValue,
		compare
	});
});
const afterRule = createRule(function after(_, options, field) {
	const compare = options.compare || "day";
	const format = options.format || field.$formats;
	const dateTime = field.$value;
	const expectedValue = typeof options.expectedValue === "function" ? options.expectedValue(field) : options.expectedValue;
	const expectedDateTime = expectedValue === "today" ? dayjs() : expectedValue === "tomorrow" ? dayjs().add(1, "day") : helpers.asDayJS(expectedValue, format).dateTime;
	if (!expectedDateTime.isValid()) throw new Error(`Invalid datetime value "${expectedValue}" provided to the after rule`);
	if (!dateTime.isAfter(expectedDateTime, compare)) field.report(messages["date.after"], "date.after", field, {
		expectedValue,
		compare
	});
});
const afterOrEqualRule = createRule(function afterOrEqual(_, options, field) {
	const compare = options.compare || "day";
	const format = options.format || field.$formats;
	const dateTime = field.$value;
	const expectedValue = typeof options.expectedValue === "function" ? options.expectedValue(field) : options.expectedValue;
	const expectedDateTime = expectedValue === "today" ? dayjs() : expectedValue === "tomorrow" ? dayjs().add(1, "day") : helpers.asDayJS(expectedValue, format).dateTime;
	if (!expectedDateTime.isValid()) throw new Error(`Invalid datetime value "${expectedValue}" provided to the afterOrEqual rule`);
	if (!dateTime.isSameOrAfter(expectedDateTime, compare)) field.report(messages["date.afterOrEqual"], "date.afterOrEqual", field, {
		expectedValue,
		compare
	});
});
const beforeRule = createRule(function before(_, options, field) {
	const compare = options.compare || "day";
	const format = options.format || field.$formats;
	const dateTime = field.$value;
	const expectedValue = typeof options.expectedValue === "function" ? options.expectedValue(field) : options.expectedValue;
	const expectedDateTime = expectedValue === "today" ? dayjs() : expectedValue === "yesterday" ? dayjs().subtract(1, "day") : helpers.asDayJS(expectedValue, format).dateTime;
	if (!expectedDateTime.isValid()) throw new Error(`Invalid datetime value "${expectedValue}" provided to the before rule`);
	if (!dateTime.isBefore(expectedDateTime, compare)) field.report(messages["date.before"], "date.before", field, {
		expectedValue,
		compare
	});
});
const beforeOrEqualRule = createRule(function beforeOrEqual(_, options, field) {
	const compare = options.compare || "day";
	const format = options.format || field.$formats;
	const dateTime = field.$value;
	const expectedValue = typeof options.expectedValue === "function" ? options.expectedValue(field) : options.expectedValue;
	const expectedDateTime = expectedValue === "today" ? dayjs() : expectedValue === "yesterday" ? dayjs().subtract(1, "day") : helpers.asDayJS(expectedValue, format).dateTime;
	if (!expectedDateTime.isValid()) throw new Error(`Invalid datetime value "${expectedValue}" provided to the beforeOrEqual rule`);
	if (!dateTime.isSameOrBefore(expectedDateTime, compare)) field.report(messages["date.beforeOrEqual"], "date.beforeOrEqual", field, {
		expectedValue,
		compare
	});
});
const sameAsRule$1 = createRule(function sameAs(_, options, field) {
	const compare = options.compare || "day";
	const dateTime = field.$value;
	const format = options.format || field.$formats;
	const expectedValue = helpers.getNestedValue(options.otherField, field);
	const expectedDateTime = helpers.asDayJS(expectedValue, format).dateTime;
	if (!expectedDateTime.isValid()) return;
	if (!dateTime.isSame(expectedDateTime, compare)) field.report(messages["date.sameAs"], "date.sameAs", field, {
		otherField: options.otherField,
		expectedValue,
		compare
	});
});
const notSameAsRule$1 = createRule(function notSameAs(_, options, field) {
	const compare = options.compare || "day";
	const dateTime = field.$value;
	const format = options.format || field.$formats;
	const expectedValue = helpers.getNestedValue(options.otherField, field);
	const expectedDateTime = helpers.asDayJS(expectedValue, format).dateTime;
	if (!expectedDateTime.isValid()) return;
	if (dateTime.isSame(expectedDateTime, compare)) field.report(messages["date.notSameAs"], "date.notSameAs", field, {
		otherField: options.otherField,
		expectedValue,
		compare
	});
});
const afterFieldRule = createRule(function afterField(_, options, field) {
	const compare = options.compare || "day";
	const dateTime = field.$value;
	const format = options.format || field.$formats;
	const expectedValue = helpers.getNestedValue(options.otherField, field);
	const expectedDateTime = helpers.asDayJS(expectedValue, format).dateTime;
	if (!expectedDateTime.isValid()) return;
	if (!dateTime.isAfter(expectedDateTime, compare)) field.report(messages["date.afterField"], "date.afterField", field, {
		otherField: options.otherField,
		expectedValue,
		compare
	});
});
const afterOrSameAsRule = createRule(function afterOrSameAs(_, options, field) {
	const compare = options.compare || "day";
	const dateTime = field.$value;
	const format = options.format || field.$formats;
	const expectedValue = helpers.getNestedValue(options.otherField, field);
	const expectedDateTime = helpers.asDayJS(expectedValue, format).dateTime;
	if (!expectedDateTime.isValid()) return;
	if (!dateTime.isSameOrAfter(expectedDateTime, compare)) field.report(messages["date.afterOrSameAs"], "date.afterOrSameAs", field, {
		otherField: options.otherField,
		expectedValue,
		compare
	});
});
const beforeFieldRule = createRule(function beforeField(_, options, field) {
	const compare = options.compare || "day";
	const dateTime = field.$value;
	const format = options.format || field.$formats;
	const expectedValue = helpers.getNestedValue(options.otherField, field);
	const expectedDateTime = helpers.asDayJS(expectedValue, format).dateTime;
	if (!expectedDateTime.isValid()) return;
	if (!dateTime.isBefore(expectedDateTime, compare)) field.report(messages["date.beforeField"], "date.beforeField", field, {
		otherField: options.otherField,
		expectedValue,
		compare
	});
});
const beforeOrSameAsRule = createRule(function beforeOrSameAs(_, options, field) {
	const compare = options.compare || "day";
	const dateTime = field.$value;
	const format = options.format || field.$formats;
	const expectedValue = helpers.getNestedValue(options.otherField, field);
	const expectedDateTime = helpers.asDayJS(expectedValue, format).dateTime;
	if (!expectedDateTime.isValid()) return;
	if (!dateTime.isSameOrBefore(expectedDateTime, compare)) field.report(messages["date.beforeOrSameAs"], "date.beforeOrSameAs", field, {
		otherField: options.otherField,
		expectedValue,
		compare
	});
});
const weekendRule = createRule(function weekend(_, __, field) {
	const day = field.$value.day();
	if (day !== 0 && day !== 6) field.report(messages["date.weekend"], "date.weekend", field);
});
const weekdayRule = createRule(function weekday(_, __, field) {
	const day = field.$value.day();
	if (day === 0 || day === 6) field.report(messages["date.weekday"], "date.weekday", field);
});
var VineDate = class VineDate extends BaseLiteralType {
	static transform(transformer) {
		globalTransforms.date = transformer;
	}
	static rules = {
		equals: equalsRule$1,
		after: afterRule,
		afterOrEqual: afterOrEqualRule,
		before: beforeRule,
		beforeOrEqual: beforeOrEqualRule,
		sameAs: sameAsRule$1,
		notSameAs: notSameAsRule$1,
		afterField: afterFieldRule,
		afterOrSameAs: afterOrSameAsRule,
		beforeField: beforeFieldRule,
		beforeOrSameAs: beforeOrSameAsRule,
		weekend: weekendRule,
		weekday: weekdayRule
	};
	[UNIQUE_NAME] = "vine.date";
	[SUBTYPE] = "date";
	[IS_OF_TYPE] = (value) => {
		if (typeof value !== "string") return false;
		return helpers.asDayJS(value, this.options.formats).dateTime.isValid();
	};
	constructor(options, validations) {
		super(options, validations || []);
		this.dataTypeValidator = dateRule(options || {});
	}
	equals(expectedValue, options) {
		return this.use(equalsRule$1({
			expectedValue,
			...options
		}));
	}
	after(expectedValue, options) {
		return this.use(afterRule({
			expectedValue,
			...options
		}));
	}
	afterOrEqual(expectedValue, options) {
		return this.use(afterOrEqualRule({
			expectedValue,
			...options
		}));
	}
	before(expectedValue, options) {
		return this.use(beforeRule({
			expectedValue,
			...options
		}));
	}
	beforeOrEqual(expectedValue, options) {
		return this.use(beforeOrEqualRule({
			expectedValue,
			...options
		}));
	}
	sameAs(otherField, options) {
		return this.use(sameAsRule$1({
			otherField,
			...options
		}));
	}
	notSameAs(otherField, options) {
		return this.use(notSameAsRule$1({
			otherField,
			...options
		}));
	}
	afterField(otherField, options) {
		return this.use(afterFieldRule({
			otherField,
			...options
		}));
	}
	afterOrSameAs(otherField, options) {
		return this.use(afterOrSameAsRule({
			otherField,
			...options
		}));
	}
	beforeField(otherField, options) {
		return this.use(beforeFieldRule({
			otherField,
			...options
		}));
	}
	beforeOrSameAs(otherField, options) {
		return this.use(beforeOrSameAsRule({
			otherField,
			...options
		}));
	}
	weekend() {
		return this.use(weekendRule());
	}
	weekday() {
		return this.use(weekdayRule());
	}
	clone() {
		return new VineDate(this.cloneOptions(), this.cloneValidations());
	}
};
var VineNull = class VineNull {
	[SUBTYPE] = "null";
	[UNIQUE_NAME] = "vine.null";
	[IS_OF_TYPE] = (value) => {
		return value === null;
	};
	options;
	constructor(options) {
		this.options = {
			bail: true,
			allowNull: true,
			isOptional: false,
			...options
		};
	}
	cloneOptions() {
		return { ...this.options };
	}
	clone() {
		return new VineNull(this.cloneOptions());
	}
	toJSONSchema() {
		return { type: "null" };
	}
	[PARSE](propertyName, refs, options) {
		return {
			type: "literal",
			subtype: this[SUBTYPE],
			fieldName: propertyName,
			propertyName: options.toCamelCase ? camelcase(propertyName) : propertyName,
			bail: this.options.bail,
			allowNull: this.options.allowNull,
			isOptional: this.options.isOptional,
			parseFnId: this.options.parse ? refs.trackParser(this.options.parse) : void 0,
			validations: []
		};
	}
};
var UnionConditional = class {
	#schema;
	#conditional;
	constructor(conditional, schema) {
		this.#schema = schema;
		this.#conditional = conditional;
	}
	toJSONSchema() {
		return this.#schema.toJSONSchema?.();
	}
	[PARSE](propertyName, refs, options) {
		return {
			conditionalFnRefId: refs.trackConditional(this.#conditional),
			schema: this.#schema[PARSE](propertyName, refs, options)
		};
	}
};
var VineOptional = class VineOptional extends ConditionalValidations {
	[SUBTYPE] = "optional";
	[UNIQUE_NAME] = "vine.optional";
	[IS_OF_TYPE] = (value) => {
		return value === null || value === void 0;
	};
	options;
	validations;
	constructor(options, validations) {
		super();
		this.options = {
			bail: true,
			allowNull: false,
			isOptional: true,
			...options
		};
		this.validations = validations || [];
	}
	cloneValidations() {
		return this.validations.map((validation) => {
			return {
				options: validation.options,
				rule: validation.rule
			};
		});
	}
	cloneOptions() {
		return { ...this.options };
	}
	compileValidations(refs) {
		return this.validations.map((validation) => {
			return {
				ruleFnId: refs.track({
					validator: validation.rule.validator,
					options: validation.options
				}),
				implicit: validation.rule.implicit,
				isAsync: validation.rule.isAsync
			};
		});
	}
	parse(callback) {
		this.options.parse = callback;
		return this;
	}
	use(validation) {
		this.validations.push(VALIDATION in validation ? validation[VALIDATION]() : validation);
		return this;
	}
	bail(state) {
		this.options.bail = state;
		return this;
	}
	clone() {
		return new VineOptional(this.cloneOptions(), this.cloneValidations());
	}
	nullable() {
		return new VineOptional({
			...this.options,
			allowNull: true
		});
	}
	toJSONSchema() {
		return {};
	}
	[PARSE](propertyName, refs, options) {
		return {
			type: "literal",
			subtype: this[SUBTYPE],
			fieldName: propertyName,
			propertyName: options.toCamelCase ? camelcase(propertyName) : propertyName,
			bail: this.options.bail,
			allowNull: this.options.allowNull,
			isOptional: this.options.isOptional,
			parseFnId: this.options.parse ? refs.trackParser(this.options.parse) : void 0,
			validations: this.compileValidations(refs)
		};
	}
};
var VineUnion = class VineUnion {
	#conditionals;
	#otherwiseCallback = (_, field) => {
		field.report(messages.union, "union", field);
	};
	constructor(conditionals) {
		this.#conditionals = conditionals;
	}
	optional() {
		const optional = new VineOptional();
		return new VineUnion([new UnionConditional(optional[IS_OF_TYPE], optional), ...this.#conditionals]);
	}
	nullable() {
		const nullable = new VineNull();
		return new VineUnion([new UnionConditional(nullable[IS_OF_TYPE], nullable), ...this.#conditionals]);
	}
	otherwise(callback) {
		this.#otherwiseCallback = callback;
		return this;
	}
	toJSONSchema() {
		return { anyOf: this.#conditionals.map((conditional) => conditional.toJSONSchema()).filter(Boolean) };
	}
	clone() {
		const cloned = new VineUnion(this.#conditionals);
		cloned.otherwise(this.#otherwiseCallback);
		return cloned;
	}
	[PARSE](propertyName, refs, options) {
		return {
			type: "union",
			fieldName: propertyName,
			propertyName: options.toCamelCase ? camelcase(propertyName) : propertyName,
			elseConditionalFnRefId: refs.trackConditional(this.#otherwiseCallback),
			conditions: this.#conditionals.map((conditional) => conditional[PARSE](propertyName, refs, options))
		};
	}
};
function union(conditionals) {
	return new VineUnion(conditionals);
}
union.if = function unionIf(conditon, schema) {
	return new UnionConditional(conditon, schema);
};
union.else = function unionElse(schema) {
	return new UnionConditional(() => true, schema);
};
var NullableModifier = class NullableModifier {
	allowNull = true;
	get isOptional() {
		return this.#parent.isOptional;
	}
	#parent;
	constructor(parent) {
		this.#parent = parent;
	}
	optional() {
		return new OptionalModifier(this);
	}
	meta(meta) {
		return new MetaModifier(this, meta);
	}
	clone() {
		return new NullableModifier(this.#parent.clone());
	}
	toJSONSchema() {
		const schema = this.#parent.toJSONSchema?.();
		if (!schema) return { type: "null" };
		if (schema.anyOf) {
			schema.anyOf.push({ type: "null" });
			return schema;
		}
		if (schema.type === void 0) {
			schema.type = "null";
			return schema;
		}
		if (typeof schema.type === "string") {
			schema.type = [schema.type, "null"];
			return schema;
		}
		if (Array.isArray(schema.type)) {
			schema.type.push("null");
			return schema;
		}
		return schema;
	}
	[PARSE](propertyName, refs, options) {
		const output = this.#parent[PARSE](propertyName, refs, options);
		if (output.type !== "union") output.allowNull = true;
		return output;
	}
};
var MetaModifier = class MetaModifier {
	get allowNull() {
		return this.#parent.allowNull;
	}
	get isOptional() {
		return this.#parent.isOptional;
	}
	#parent;
	#meta;
	constructor(parent, meta) {
		this.#parent = parent;
		this.#meta = meta;
	}
	clone() {
		return new MetaModifier(this.#parent.clone(), this.#meta);
	}
	toJSONSchema() {
		return {
			...this.#parent.toJSONSchema?.() ?? {},
			...this.#meta
		};
	}
	[PARSE](propertyName, refs, options) {
		return this.#parent[PARSE](propertyName, refs, options);
	}
};
var OptionalModifier = class OptionalModifier extends ConditionalValidations {
	isOptional = true;
	get allowNull() {
		return this.#parent.allowNull;
	}
	#parent;
	validations;
	constructor(parent, validations) {
		super();
		this.#parent = parent;
		this.validations = validations || [];
	}
	cloneValidations() {
		return this.validations.map((validation) => {
			return {
				options: validation.options,
				rule: validation.rule
			};
		});
	}
	compileValidations(refs) {
		return this.validations.map((validation) => {
			return {
				ruleFnId: refs.track({
					validator: validation.rule.validator,
					options: validation.options
				}),
				name: validation.rule.name,
				implicit: validation.rule.implicit,
				isAsync: validation.rule.isAsync
			};
		});
	}
	nullable() {
		return new NullableModifier(this);
	}
	meta(meta) {
		return new MetaModifier(this, meta);
	}
	use(validation) {
		this.validations.push(VALIDATION in validation ? validation[VALIDATION]() : validation);
		return this;
	}
	clone() {
		return new OptionalModifier(this.#parent.clone(), this.cloneValidations());
	}
	toJSONSchema() {
		return { ...this.#parent.toJSONSchema() };
	}
	[PARSE](propertyName, refs, options) {
		const output = this.#parent[PARSE](propertyName, refs, options);
		if (output.type !== "union") {
			output.isOptional = true;
			output.validations = output.validations.concat(this.compileValidations(refs));
		}
		return output;
	}
};
var BaseType = class extends Macroable {
	toJSONSchema() {
		return {};
	}
	validations;
	options;
	constructor(options, validations) {
		super();
		this.options = options || {
			bail: true,
			allowNull: false,
			isOptional: false
		};
		this.validations = validations || [];
	}
	cloneValidations() {
		return this.validations.map((validation) => {
			return {
				options: validation.options,
				rule: validation.rule
			};
		});
	}
	cloneOptions() {
		return { ...this.options };
	}
	compileValidations(refs) {
		return this.validations.map((validation) => {
			return {
				ruleFnId: refs.track({
					validator: validation.rule.validator,
					options: validation.options
				}),
				name: validation.rule.name,
				implicit: validation.rule.implicit,
				isAsync: validation.rule.isAsync
			};
		});
	}
	parse(callback) {
		this.options.parse = callback;
		return this;
	}
	use(validation) {
		this.validations.push(VALIDATION in validation ? validation[VALIDATION]() : validation);
		return this;
	}
	bail(state) {
		this.options.bail = state;
		return this;
	}
	optional() {
		return new OptionalModifier(this);
	}
	nullable() {
		return new NullableModifier(this);
	}
	meta(meta) {
		return new MetaModifier(this, meta);
	}
};
var VineTuple = class VineTuple extends BaseType {
	#schemas;
	#allowUnknownProperties = false;
	[UNIQUE_NAME] = "vine.array";
	[IS_OF_TYPE] = (value) => {
		return Array.isArray(value);
	};
	constructor(schemas, options, validations) {
		super(options, validations);
		this.#schemas = schemas;
	}
	allowUnknownProperties() {
		this.#allowUnknownProperties = true;
		return this;
	}
	clone() {
		const cloned = new VineTuple(this.#schemas.map((schema) => schema.clone()), this.cloneOptions(), this.cloneValidations());
		if (this.#allowUnknownProperties) cloned.allowUnknownProperties();
		return cloned;
	}
	toJSONSchema() {
		const items = [];
		for (const item of this.#schemas) {
			if (!item.toJSONSchema) continue;
			items.push(item.toJSONSchema());
		}
		const schema = {
			type: "array",
			minItems: this.#schemas.length,
			maxItems: this.#schemas.length,
			additionalItems: false,
			items
		};
		for (const validation of this.validations) {
			if (!validation.rule.toJSONSchema) continue;
			validation.rule.toJSONSchema(schema, validation.options);
		}
		return schema;
	}
	[PARSE](propertyName, refs, options) {
		return {
			type: "tuple",
			fieldName: propertyName,
			propertyName: options.toCamelCase ? camelcase(propertyName) : propertyName,
			bail: this.options.bail,
			allowNull: this.options.allowNull,
			isOptional: this.options.isOptional,
			allowUnknownProperties: this.#allowUnknownProperties,
			parseFnId: this.options.parse ? refs.trackParser(this.options.parse) : void 0,
			validations: this.compileValidations(refs),
			properties: this.#schemas.map((schema, index) => schema[PARSE](String(index), refs, options))
		};
	}
};
const minLengthRule$2 = createRule(function minLength(value, options, field) {
	if (value.length < options.min) field.report(messages["array.minLength"], "array.minLength", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.minItems = options.min;
} });
const maxLengthRule$2 = createRule(function maxLength(value, options, field) {
	if (value.length > options.max) field.report(messages["array.maxLength"], "array.maxLength", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.maxItems = options.max;
} });
const fixedLengthRule$2 = createRule(function fixedLength(value, options, field) {
	if (value.length !== options.size) field.report(messages["array.fixedLength"], "array.fixedLength", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.minItems = options.size;
	schema.maxItems = options.size;
} });
const notEmptyRule = createRule(function notEmpty(value, _, field) {
	if (value.length <= 0) field.report(messages.notEmpty, "notEmpty", field);
}, { toJSONSchema: (schema) => {
	schema.minItems = 1;
} });
const distinctRule = createRule(function distinct(value, options, field) {
	if (!helpers.isDistinct(value, options.fields)) field.report(messages.distinct, "distinct", field, options);
}, { toJSONSchema: (schema) => {
	schema.uniqueItems = true;
} });
const compactRule = createRule(function compact(value, _, field) {
	if (!field.isValid) return;
	field.mutate(value.filter((item) => helpers.exists(item) && item !== ""), field);
});
var VineArray = class VineArray extends BaseType {
	static rules = {
		compact: compactRule,
		notEmpty: notEmptyRule,
		distinct: distinctRule,
		minLength: minLengthRule$2,
		maxLength: maxLengthRule$2,
		fixedLength: fixedLengthRule$2
	};
	#schema;
	[UNIQUE_NAME] = "vine.array";
	[IS_OF_TYPE] = (value) => {
		return Array.isArray(value);
	};
	constructor(schema, options, validations) {
		super(options, validations);
		this.#schema = schema;
	}
	minLength(expectedLength) {
		return this.use(minLengthRule$2({ min: expectedLength }));
	}
	maxLength(expectedLength) {
		return this.use(maxLengthRule$2({ max: expectedLength }));
	}
	fixedLength(expectedLength) {
		return this.use(fixedLengthRule$2({ size: expectedLength }));
	}
	notEmpty() {
		return this.use(notEmptyRule());
	}
	distinct(fields) {
		return this.use(distinctRule({ fields }));
	}
	compact() {
		return this.use(compactRule());
	}
	clone() {
		return new VineArray(this.#schema.clone(), this.cloneOptions(), this.cloneValidations());
	}
	toJSONSchema() {
		const schema = {
			type: "array",
			items: this.#schema.toJSONSchema?.() ?? {}
		};
		for (const validation of this.validations) {
			if (!validation.rule.toJSONSchema) continue;
			validation.rule.toJSONSchema(schema, validation.options);
		}
		return schema;
	}
	[PARSE](propertyName, refs, options) {
		return {
			type: "array",
			fieldName: propertyName,
			propertyName: options.toCamelCase ? camelcase(propertyName) : propertyName,
			bail: this.options.bail,
			allowNull: this.options.allowNull,
			isOptional: this.options.isOptional,
			each: this.#schema[PARSE]("*", refs, options),
			parseFnId: this.options.parse ? refs.trackParser(this.options.parse) : void 0,
			validations: this.compileValidations(refs)
		};
	}
};
var VineCamelCaseObject = class VineCamelCaseObject extends BaseType {
	#schema;
	[UNIQUE_NAME] = "types.object";
	[IS_OF_TYPE] = (value) => {
		return value !== null && typeof value === "object" && !Array.isArray(value);
	};
	constructor(schema) {
		super();
		this.#schema = schema;
	}
	clone() {
		return new VineCamelCaseObject(this.#schema.clone());
	}
	toJSONSchema() {
		return this.#schema.toJSONSchema();
	}
	[PARSE](propertyName, refs, options) {
		options.toCamelCase = true;
		return this.#schema[PARSE](propertyName, refs, options);
	}
};
var VineObject = class VineObject extends BaseType {
	#properties;
	#groups = [];
	#allowUnknownProperties = false;
	[UNIQUE_NAME] = "vine.object";
	[IS_OF_TYPE] = (value) => {
		return value !== null && typeof value === "object" && !Array.isArray(value);
	};
	constructor(properties, options, validations) {
		if (!properties) throw new Error("Missing properties for \"vine.object\". Use an empty object if you do not want to validate any specific fields");
		super(options, validations);
		this.#properties = properties;
	}
	getProperties() {
		return Object.keys(this.#properties).reduce((result, key) => {
			result[key] = this.#properties[key].clone();
			return result;
		}, {});
	}
	pick(keys) {
		const result = {};
		for (const key of keys) result[key] = this.#properties[key].clone();
		return result;
	}
	omit(keys) {
		const result = {};
		for (const key of Object.keys(this.#properties)) if (!keys.includes(key)) result[key] = this.#properties[key].clone();
		return result;
	}
	allowUnknownProperties() {
		this.#allowUnknownProperties = true;
		return this;
	}
	merge(group) {
		this.#groups.push(group);
		return this;
	}
	clone() {
		const cloned = new VineObject(this.getProperties(), this.cloneOptions(), this.cloneValidations());
		this.#groups.forEach((group) => cloned.merge(group));
		if (this.#allowUnknownProperties) cloned.allowUnknownProperties();
		return cloned;
	}
	toCamelCase() {
		return new VineCamelCaseObject(this);
	}
	toJSONSchema() {
		const properties = {};
		const required = [];
		for (const [key, property] of Object.entries(this.getProperties())) {
			if (!property.toJSONSchema) continue;
			properties[key] = property.toJSONSchema();
			if (property.isOptional !== true && property.allowNull !== true) required.push(key);
		}
		const schema = {
			type: "object",
			properties,
			required,
			additionalProperties: this.#allowUnknownProperties
		};
		for (const validation of this.validations) {
			if (!validation.rule.toJSONSchema) continue;
			validation.rule.toJSONSchema(schema, validation.options);
		}
		if (this.#groups.length > 0) return { anyOf: [...this.#groups.map((group) => group.toJSONSchema()), schema] };
		return schema;
	}
	partial(keys) {
		if (this.#groups.length > 0 || this.#allowUnknownProperties) throw new Error("toOptional cannot be used on schemas that have groups or allowUnknownProperties enabled");
		const properties = {};
		for (const key of Object.keys(this.#properties)) {
			let field = this.#properties[key].clone();
			if ((!keys || keys.includes(key)) && "optional" in field && typeof field.optional === "function") field = field.optional();
			properties[key] = field;
		}
		return new VineObject(properties, this.cloneOptions(), this.cloneValidations());
	}
	[PARSE](propertyName, refs, options) {
		return {
			type: "object",
			fieldName: propertyName,
			propertyName: options.toCamelCase ? camelcase(propertyName) : propertyName,
			bail: this.options.bail,
			allowNull: this.options.allowNull,
			isOptional: this.options.isOptional,
			parseFnId: this.options.parse ? refs.trackParser(this.options.parse) : void 0,
			allowUnknownProperties: this.#allowUnknownProperties,
			validations: this.compileValidations(refs),
			properties: Object.keys(this.#properties).map((property) => {
				return this.#properties[property][PARSE](property, refs, options);
			}),
			groups: this.#groups.map((group) => {
				return group[PARSE](refs, options);
			})
		};
	}
};
const minLengthRule$1 = createRule(function minLength(value, options, field) {
	if (Object.keys(value).length < options.min) field.report(messages["record.minLength"], "record.minLength", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.minProperties = options.min;
} });
const maxLengthRule$1 = createRule(function maxLength(value, options, field) {
	if (Object.keys(value).length > options.max) field.report(messages["record.maxLength"], "record.maxLength", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.maxProperties = options.max;
} });
const fixedLengthRule$1 = createRule(function fixedLength(value, options, field) {
	if (Object.keys(value).length !== options.size) field.report(messages["record.fixedLength"], "record.fixedLength", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.minProperties = options.size;
	schema.maxProperties = options.size;
} });
const validateKeysRule = createRule(function validateKeys(value, callback, field) {
	callback(Object.keys(value), field);
});
var VineRecord = class VineRecord extends BaseType {
	static rules = {
		maxLength: maxLengthRule$1,
		minLength: minLengthRule$1,
		fixedLength: fixedLengthRule$1,
		validateKeys: validateKeysRule
	};
	#schema;
	[UNIQUE_NAME] = "vine.object";
	[IS_OF_TYPE] = (value) => {
		return value !== null && typeof value === "object" && !Array.isArray(value);
	};
	constructor(schema, options, validations) {
		super(options, validations);
		this.#schema = schema;
	}
	minLength(expectedLength) {
		return this.use(minLengthRule$1({ min: expectedLength }));
	}
	maxLength(expectedLength) {
		return this.use(maxLengthRule$1({ max: expectedLength }));
	}
	fixedLength(expectedLength) {
		return this.use(fixedLengthRule$1({ size: expectedLength }));
	}
	validateKeys(...args) {
		return this.use(validateKeysRule(...args));
	}
	clone() {
		return new VineRecord(this.#schema.clone(), this.cloneOptions(), this.cloneValidations());
	}
	toJSONSchema() {
		const schema = {
			type: "object",
			additionalProperties: {}
		};
		schema.additionalProperties = this.#schema.toJSONSchema?.() ?? {};
		for (const validation of this.validations) {
			if (!validation.rule.toJSONSchema) continue;
			validation.rule.toJSONSchema(schema, validation.options);
		}
		return schema;
	}
	[PARSE](propertyName, refs, options) {
		return {
			type: "record",
			fieldName: propertyName,
			propertyName: options.toCamelCase ? camelcase(propertyName) : propertyName,
			bail: this.options.bail,
			allowNull: this.options.allowNull,
			isOptional: this.options.isOptional,
			each: this.#schema[PARSE]("*", refs, options),
			parseFnId: this.options.parse ? refs.trackParser(this.options.parse) : void 0,
			validations: this.compileValidations(refs)
		};
	}
};
const stringRule = createRule(function string(value, _, field) {
	if (!field.isDefined) return false;
	if (typeof value === "string") return true;
	field.report(messages.string, "string", field);
	return false;
}, { toJSONSchema: (schema) => {
	schema.type = "string";
} });
const emailRule = createRule(function email(value, options, field) {
	if (!helpers.isEmail(value, options)) field.report(messages.email, "email", field);
}, { toJSONSchema: (schema) => {
	schema.format = "email";
} });
const mobileRule = createRule(function mobile(value, options, field) {
	const normalizedOptions = options && typeof options === "function" ? options(field) : options;
	const locales = normalizedOptions?.locale || "any";
	if (!helpers.isMobilePhone(value, locales, normalizedOptions)) field.report(messages.mobile, "mobile", field);
});
const ipAddressRule = createRule(function ipAddress(value, options, field) {
	if (!helpers.isIP(value, options?.version)) field.report(messages.ipAddress, "ipAddress", field);
}, { toJSONSchema: (schema, options) => {
	schema.format = options?.version === 6 ? "ipv6" : "ipv4";
} });
const regexRule = createRule(function regex(value, expression, field) {
	if (!expression.test(value)) field.report(messages.regex, "regex", field);
}, { toJSONSchema: (schema, options) => {
	schema.pattern = options.source;
} });
const hexCodeRule = createRule(function hexCode(value, _, field) {
	if (!helpers.isHexColor(value)) field.report(messages.hexCode, "hexCode", field);
}, { toJSONSchema: (schema) => {
	schema.pattern = "^#?([0-9a-f]{6}|[0-9a-f]{3}|[0-9a-f]{8})$";
} });
const urlRule = createRule(function url(value, options, field) {
	if (!helpers.isURL(value, options)) field.report(messages.url, "url", field);
}, { toJSONSchema: (schema) => {
	schema.format = "uri";
} });
const activeUrlRule = createRule(async function activeUrl(value, _, field) {
	if (!await helpers.isActiveURL(value)) field.report(messages.activeUrl, "activeUrl", field);
});
const alphaRule = createRule(function alpha(value, options, field) {
	let characterSet = "a-zA-Z";
	if (options) {
		if (options.allowSpaces) characterSet += "\\s";
		if (options.allowDashes) characterSet += "-";
		if (options.allowUnderscores) characterSet += "_";
	}
	if (!new RegExp(`^[${characterSet}]+$`).test(value)) field.report(messages.alpha, "alpha", field);
}, { toJSONSchema: (schema, options) => {
	let characterSet = "a-zA-Z";
	if (options) {
		if (options.allowSpaces) characterSet += "\\s";
		if (options.allowDashes) characterSet += "-";
		if (options.allowUnderscores) characterSet += "_";
	}
	schema.pattern = `^[${characterSet}]+$`;
} });
const alphaNumericRule = createRule(function alphaNumeric(value, options, field) {
	let characterSet = "a-zA-Z0-9";
	if (options) {
		if (options.allowSpaces) characterSet += "\\s";
		if (options.allowDashes) characterSet += "-";
		if (options.allowUnderscores) characterSet += "_";
	}
	if (!new RegExp(`^[${characterSet}]+$`).test(value)) field.report(messages.alphaNumeric, "alphaNumeric", field);
}, { toJSONSchema: (schema, options) => {
	let characterSet = "a-zA-Z0-9";
	if (options) {
		if (options.allowSpaces) characterSet += "\\s";
		if (options.allowDashes) characterSet += "-";
		if (options.allowUnderscores) characterSet += "_";
	}
	schema.pattern = `^[${characterSet}]+$`;
} });
const minLengthRule = createRule(function minLength(value, options, field) {
	if (value.length < options.min) field.report(messages.minLength, "minLength", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.minLength = options.min;
} });
const maxLengthRule = createRule(function maxLength(value, options, field) {
	if (value.length > options.max) field.report(messages.maxLength, "maxLength", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.maxLength = options.max;
} });
const fixedLengthRule = createRule(function fixedLength(value, options, field) {
	if (value.length !== options.size) field.report(messages.fixedLength, "fixedLength", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.minLength = options.size;
	schema.maxLength = options.size;
} });
const endsWithRule = createRule(function endsWith(value, options, field) {
	if (!value.endsWith(options.substring)) field.report(messages.endsWith, "endsWith", field, options);
});
const startsWithRule = createRule(function startsWith(value, options, field) {
	if (!value.startsWith(options.substring)) field.report(messages.startsWith, "startsWith", field, options);
});
const sameAsRule = createRule(function sameAs(value, options, field) {
	if (helpers.getNestedValue(options.otherField, field) !== value) {
		field.report(messages.sameAs, "sameAs", field, options);
		return;
	}
});
const notSameAsRule = createRule(function notSameAs(value, options, field) {
	if (helpers.getNestedValue(options.otherField, field) === value) {
		field.report(messages.notSameAs, "notSameAs", field, options);
		return;
	}
});
const confirmedRule = createRule(function confirmed(value, options, field) {
	const normalizedOptions = options ?? {};
	const otherField = normalizedOptions.as ?? normalizedOptions.confirmationField ?? `${field.name}_confirmation`;
	const input = field.parent[otherField];
	if (input !== value) {
		field.report(messages.confirmed, "confirmed", {
			...field,
			name: otherField,
			wildCardPath: `${field.wildCardPath.replace(String(field.name), otherField)}`,
			isDefined: true,
			isValid: false,
			value: input,
			getFieldPath() {
				return `${field.getFieldPath().replace(String(field.name), otherField)}`;
			}
		}, {
			otherField,
			originalField: field.name
		});
		return;
	}
});
const inRule$1 = createRule(function inList(value, options, field) {
	if (!(typeof options.choices === "function" ? options.choices(field) : options.choices).includes(value)) {
		field.report(messages.in, "in", field, options);
		return;
	}
});
const notInRule = createRule(function notIn(value, options, field) {
	if ((typeof options.list === "function" ? options.list(field) : options.list).includes(value)) {
		field.report(messages.notIn, "notIn", field, options);
		return;
	}
});
const creditCardRule = createRule(function creditCard(value, options, field) {
	const providers = options ? typeof options === "function" ? options(field)?.provider || [] : options.provider : [];
	if (!providers.length) {
		if (!helpers.isCreditCard(value)) field.report(messages.creditCard, "creditCard", field, { providersList: "credit" });
	} else if (!providers.find((provider) => helpers.isCreditCard(value, { provider }))) field.report(messages.creditCard, "creditCard", field, {
		providers,
		providersList: providers.join("/")
	});
});
const passportRule = createRule(function passport(value, options, field) {
	const countryCodes = typeof options === "function" ? options(field).countryCode : options.countryCode;
	if (!countryCodes.find((countryCode) => helpers.isPassportNumber(value, countryCode))) field.report(messages.passport, "passport", field, { countryCodes });
});
const vatRule = createRule(function vat(value, options, field) {
	const countryCodes = typeof options === "function" ? options(field).countryCode : options.countryCode;
	if (!countryCodes.find((countryCode) => helpers.isVAT(value, countryCode))) field.report(messages.vat, "vat", field, { countryCodes });
});
const postalCodeRule = createRule(function postalCode(value, options, field) {
	const countryCodes = options ? typeof options === "function" ? options(field)?.countryCode || [] : options.countryCode : [];
	if (!countryCodes.length) {
		if (!helpers.isPostalCode(value, "any")) field.report(messages.postalCode, "postalCode", field);
	} else if (!countryCodes.find((countryCode) => helpers.isPostalCode(value, countryCode))) field.report(messages.postalCode, "postalCode", field, { countryCodes });
});
const uuidRule = createRule(function uuid(value, options, field) {
	if (!options || !options.version) {
		if (!helpers.isUUID(value)) field.report(messages.uuid, "uuid", field);
	} else if (!options.version.find((version) => helpers.isUUID(value, version))) field.report(messages.uuid, "uuid", field, options);
}, { toJSONSchema: (schema) => {
	schema.format = "uuid";
} });
const ulidRule = createRule(function ulid(value, _, field) {
	if (!helpers.isULID(value)) field.report(messages.ulid, "ulid", field);
}, { toJSONSchema: (schema) => {
	schema.pattern = "^[0-7][0-9A-HJKMNP-TV-Z]{25}$";
} });
const asciiRule = createRule(function ascii(value, _, field) {
	if (!helpers.isAscii(value)) field.report(messages.ascii, "ascii", field);
});
const ibanRule = createRule(function iban(value, _, field) {
	if (!helpers.isIBAN(value)) field.report(messages.iban, "iban", field);
});
const jwtRule = createRule(function jwt(value, _, field) {
	if (!helpers.isJWT(value)) field.report(messages.jwt, "jwt", field);
});
const coordinatesRule = createRule(function coordinates(value, _, field) {
	if (!helpers.isLatLong(value)) field.report(messages.coordinates, "coordinates", field);
});
const trimRule = createRule(function trim(value, _, field) {
	if (!field.isValid) return;
	field.mutate(value.trim(), field);
});
const normalizeEmailRule = createRule(function normalizeEmail(value, options, field) {
	if (!field.isValid) return;
	field.mutate(normalizeEmailValue.default(value, options), field);
});
const toUpperCaseRule = createRule(function toUpperCase(value, locales, field) {
	if (!field.isValid) return;
	field.mutate(value.toLocaleUpperCase(locales), field);
});
const toLowerCaseRule = createRule(function toLowerCase(value, locales, field) {
	if (!field.isValid) return;
	field.mutate(value.toLocaleLowerCase(locales), field);
});
const toCamelCaseRule = createRule(function toCamelCase(value, _, field) {
	if (!field.isValid) return;
	field.mutate(camelcase(value), field);
});
const escapeRule = createRule(function escape(value, _, field) {
	if (!field.isValid) return;
	field.mutate(escapeValue.default(value), field);
});
const normalizeUrlRule = createRule(function normalizeUrlValue(value, options, field) {
	if (!field.isValid) return;
	field.mutate(normalizeUrl(value, options), field);
});
var VineString = class VineString extends BaseLiteralType {
	static rules = {
		in: inRule$1,
		jwt: jwtRule,
		url: urlRule,
		iban: ibanRule,
		uuid: uuidRule,
		ulid: ulidRule,
		trim: trimRule,
		email: emailRule,
		alpha: alphaRule,
		ascii: asciiRule,
		notIn: notInRule,
		regex: regexRule,
		escape: escapeRule,
		sameAs: sameAsRule,
		mobile: mobileRule,
		string: stringRule,
		hexCode: hexCodeRule,
		passport: passportRule,
		endsWith: endsWithRule,
		confirmed: confirmedRule,
		activeUrl: activeUrlRule,
		minLength: minLengthRule,
		notSameAs: notSameAsRule,
		maxLength: maxLengthRule,
		vat: vatRule,
		ipAddress: ipAddressRule,
		creditCard: creditCardRule,
		postalCode: postalCodeRule,
		startsWith: startsWithRule,
		toUpperCase: toUpperCaseRule,
		toLowerCase: toLowerCaseRule,
		toCamelCase: toCamelCaseRule,
		fixedLength: fixedLengthRule,
		coordinates: coordinatesRule,
		normalizeUrl: normalizeUrlRule,
		alphaNumeric: alphaNumericRule,
		normalizeEmail: normalizeEmailRule
	};
	[SUBTYPE] = "string";
	[UNIQUE_NAME] = "vine.string";
	[IS_OF_TYPE] = (value) => {
		return typeof value === "string";
	};
	constructor(options, validations) {
		super(options, validations || []);
		this.dataTypeValidator = stringRule();
	}
	url(...args) {
		return this.use(urlRule(...args));
	}
	activeUrl() {
		return this.use(activeUrlRule());
	}
	email(...args) {
		return this.use(emailRule(...args));
	}
	mobile(...args) {
		return this.use(mobileRule(...args));
	}
	vat(...args) {
		return this.use(vatRule(...args));
	}
	ipAddress(version) {
		return this.use(ipAddressRule(version ? { version } : void 0));
	}
	hexCode() {
		return this.use(hexCodeRule());
	}
	regex(expression) {
		return this.use(regexRule(expression));
	}
	alpha(options) {
		return this.use(alphaRule(options));
	}
	alphaNumeric(options) {
		return this.use(alphaNumericRule(options));
	}
	minLength(expectedLength) {
		return this.use(minLengthRule({ min: expectedLength }));
	}
	maxLength(expectedLength) {
		return this.use(maxLengthRule({ max: expectedLength }));
	}
	fixedLength(expectedLength) {
		return this.use(fixedLengthRule({ size: expectedLength }));
	}
	confirmed(options) {
		return this.use(confirmedRule(options));
	}
	trim() {
		return this.use(trimRule());
	}
	normalizeEmail(options) {
		return this.use(normalizeEmailRule(options));
	}
	toUpperCase() {
		return this.use(toUpperCaseRule());
	}
	toLowerCase() {
		return this.use(toLowerCaseRule());
	}
	toCamelCase() {
		return this.use(toCamelCaseRule());
	}
	escape() {
		return this.use(escapeRule());
	}
	normalizeUrl(...args) {
		return this.use(normalizeUrlRule(...args));
	}
	startsWith(substring) {
		return this.use(startsWithRule({ substring }));
	}
	endsWith(substring) {
		return this.use(endsWithRule({ substring }));
	}
	sameAs(otherField) {
		return this.use(sameAsRule({ otherField }));
	}
	notSameAs(otherField) {
		return this.use(notSameAsRule({ otherField }));
	}
	in(choices) {
		return this.use(inRule$1({ choices }));
	}
	notIn(list) {
		return this.use(notInRule({ list }));
	}
	creditCard(...args) {
		return this.use(creditCardRule(...args));
	}
	passport(...args) {
		return this.use(passportRule(...args));
	}
	postalCode(...args) {
		return this.use(postalCodeRule(...args));
	}
	uuid(...args) {
		return this.use(uuidRule(...args));
	}
	ulid() {
		return this.use(ulidRule());
	}
	ascii() {
		return this.use(asciiRule());
	}
	iban() {
		return this.use(ibanRule());
	}
	jwt() {
		return this.use(jwtRule());
	}
	coordinates() {
		return this.use(coordinatesRule());
	}
	clone() {
		return new VineString(this.cloneOptions(), this.cloneValidations());
	}
};
const numberRule = createRule(function number(value, options, field) {
	if (!field.isDefined) return false;
	const valueAsNumber = options.strict ? value : helpers.asNumber(value);
	if (typeof valueAsNumber !== "number" || Number.isNaN(valueAsNumber) || valueAsNumber === Number.POSITIVE_INFINITY || valueAsNumber === Number.NEGATIVE_INFINITY) {
		field.report(messages.number, "number", field);
		return false;
	}
	field.mutate(valueAsNumber, field);
	return true;
}, { toJSONSchema: (schema) => {
	schema.type = "number";
} });
const minRule = createRule(function min(value, options, field) {
	if (value < options.min) field.report(messages.min, "min", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.minimum = options.min;
} });
const maxRule = createRule(function max(value, options, field) {
	if (value > options.max) field.report(messages.max, "max", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.maximum = options.max;
} });
const rangeRule = createRule(function range(value, options, field) {
	if (value < options.min || value > options.max) field.report(messages.range, "range", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.minimum = options.min;
	schema.maximum = options.max;
} });
const positiveRule = createRule(function positive(value, _, field) {
	if (value > 0) return;
	field.report(messages.positive, "positive", field);
}, { toJSONSchema: (schema) => {
	schema.minimum = 0;
} });
const negativeRule = createRule(function negative(value, _, field) {
	if (value < 0) return;
	field.report(messages.negative, "negative", field);
}, { toJSONSchema: (schema) => {
	schema.exclusiveMaximum = 0;
} });
const nonNegativeRule = createRule(function nonNegative(value, _, field) {
	if (value >= 0) return;
	field.report(messages.nonNegative, "nonNegative", field);
});
const nonPositiveRule = createRule(function nonPositive(value, _, field) {
	if (value <= 0) return;
	field.report(messages.nonPositive, "nonPositive", field);
});
const decimalRule = createRule(function decimal(value, options, field) {
	if (!helpers.isDecimal(String(value), {
		force_decimal: options.range[0] !== 0,
		decimal_digits: options.range.join(",")
	})) field.report(messages.decimal, "decimal", field, { digits: options.range.join("-") });
});
const withoutDecimalsRule = createRule(function withoutDecimals(value, _, field) {
	if (!Number.isInteger(value)) field.report(messages.withoutDecimals, "withoutDecimals", field);
}, { toJSONSchema: (schema) => {
	schema.type = "integer";
} });
const inRule = createRule(function inValues(value, options, field) {
	if (!options.values.includes(value)) field.report(messages["number.in"], "in", field, options);
}, { toJSONSchema: (schema, options) => {
	schema.enum = options.values;
} });
var VineNumber = class VineNumber extends BaseLiteralType {
	static rules = {
		in: inRule,
		max: maxRule,
		min: minRule,
		range: rangeRule,
		number: numberRule,
		decimal: decimalRule,
		negative: negativeRule,
		positive: positiveRule,
		nonNegativeRule,
		nonPositiveRule,
		withoutDecimals: withoutDecimalsRule
	};
	[SUBTYPE] = "number";
	[UNIQUE_NAME] = "vine.number";
	[IS_OF_TYPE] = (value) => {
		const valueAsNumber = helpers.asNumber(value);
		return !Number.isNaN(valueAsNumber);
	};
	constructor(options, validations) {
		super(options, validations || []);
		this.dataTypeValidator = numberRule(options || {});
	}
	min(value) {
		return this.use(minRule({ min: value }));
	}
	max(value) {
		return this.use(maxRule({ max: value }));
	}
	range(value) {
		return this.use(rangeRule({
			min: value[0],
			max: value[1]
		}));
	}
	positive() {
		return this.use(positiveRule());
	}
	negative() {
		return this.use(negativeRule());
	}
	nonNegative() {
		return this.use(nonNegativeRule());
	}
	nonPositive() {
		return this.use(nonPositiveRule());
	}
	decimal(range) {
		return this.use(decimalRule({ range: Array.isArray(range) ? range : [range] }));
	}
	withoutDecimals() {
		return this.use(withoutDecimalsRule());
	}
	clone() {
		return new VineNumber(this.cloneOptions(), this.cloneValidations());
	}
	in(values) {
		return this.use(inRule({ values }));
	}
};
const booleanRule = createRule(function boolean(value, options, field) {
	const valueAsBoolean = options.strict === true ? value : helpers.asBoolean(value);
	if (typeof valueAsBoolean !== "boolean") {
		field.report(messages.boolean, "boolean", field);
		return;
	}
	field.mutate(valueAsBoolean, field);
}, { toJSONSchema: (schema, { strict = false }) => {
	if (strict) schema.type = "boolean";
	else schema.enum = [...BOOLEAN_POSITIVES, ...BOOLEAN_NEGATIVES];
} });
var VineBoolean = class VineBoolean extends BaseLiteralType {
	static rules = { boolean: booleanRule };
	[SUBTYPE] = "boolean";
	[UNIQUE_NAME] = "vine.boolean";
	[IS_OF_TYPE] = (value) => {
		return typeof (this.options.strict === true ? value : helpers.asBoolean(value)) === "boolean";
	};
	constructor(options, validations) {
		super(options, validations || [booleanRule(options || {})]);
	}
	clone() {
		return new VineBoolean(this.cloneOptions(), this.cloneValidations());
	}
};
const equalsRule = createRule(function equals(value, options, field) {
	const comparedValue = helpers.compareValues(value, options.expectedValue);
	if (!comparedValue.isEqual) {
		field.report(messages.literal, "literal", field, options);
		return;
	}
	field.mutate(comparedValue.casted, field);
});
var VineLiteral = class VineLiteral extends BaseLiteralType {
	static rules = { equals: equalsRule };
	#value;
	[SUBTYPE] = "literal";
	[IS_OF_TYPE] = (value) => {
		return helpers.compareValues(value, this.#value).isEqual;
	};
	constructor(value, options, validations) {
		super(options, validations || [equalsRule({ expectedValue: value })]);
		this.#value = value;
		this[UNIQUE_NAME] = `vine.literal.${this.#value}`;
	}
	clone() {
		return new VineLiteral(this.#value, this.cloneOptions(), this.cloneValidations());
	}
	toJSONSchema() {
		const schema = super.toJSONSchema();
		if (typeof this.#value === "string") {
			schema.type = "string";
			schema.enum = [this.#value];
		}
		if (typeof this.#value === "boolean") {
			schema.type = "boolean";
			schema.enum = [this.#value];
		}
		if (typeof this.#value === "number") {
			schema.type = "number";
			schema.enum = [this.#value];
		}
		return schema;
	}
};
const ACCEPTED_VALUES = [
	"on",
	"1",
	"yes",
	"true",
	true,
	1
];
const acceptedRule = createRule(function accepted(value, _, field) {
	if (!ACCEPTED_VALUES.includes(value)) field.report(messages.accepted, "accepted", field);
});
var VineAccepted = class VineAccepted extends BaseLiteralType {
	static rules = { accepted: acceptedRule };
	[SUBTYPE] = "checkbox";
	constructor(options, validations) {
		super(options, validations || [acceptedRule()]);
	}
	clone() {
		return new VineAccepted(this.cloneOptions(), this.cloneValidations());
	}
};
var ObjectGroup = class ObjectGroup {
	#conditionals;
	#otherwiseCallback = (_, field) => {
		field.report(messages.unionGroup, "unionGroup", field);
	};
	constructor(conditionals) {
		this.#conditionals = conditionals;
	}
	toJSONSchema() {
		return { anyOf: this.#conditionals.map((conditional) => conditional.toJSONSchema()) };
	}
	clone() {
		const cloned = new ObjectGroup(this.#conditionals);
		cloned.otherwise(this.#otherwiseCallback);
		return cloned;
	}
	otherwise(callback) {
		this.#otherwiseCallback = callback;
		return this;
	}
	[PARSE](refs, options) {
		return {
			type: "group",
			elseConditionalFnRefId: refs.trackConditional(this.#otherwiseCallback),
			conditions: this.#conditionals.map((conditional) => conditional[PARSE](refs, options))
		};
	}
};
var GroupConditional = class {
	#properties;
	#conditional;
	constructor(conditional, properties) {
		this.#properties = properties;
		this.#conditional = conditional;
	}
	toJSONSchema() {
		const properties = {};
		const required = [];
		for (const [key, property] of Object.entries(this.#properties)) {
			properties[key] = property.toJSONSchema();
			if (!("isOptional" in property) || property.isOptional !== true) required.push(key);
		}
		return {
			type: "object",
			properties,
			required
		};
	}
	[PARSE](refs, options) {
		return {
			schema: {
				type: "sub_object",
				properties: Object.keys(this.#properties).map((property) => {
					return this.#properties[property][PARSE](property, refs, options);
				}),
				groups: []
			},
			conditionalFnRefId: refs.trackConditional(this.#conditional)
		};
	}
};
function group(conditionals) {
	return new ObjectGroup(conditionals);
}
group.if = function groupIf(conditon, properties) {
	return new GroupConditional(conditon, properties);
};
group.else = function groupElse(properties) {
	return new GroupConditional(() => true, properties);
};
var VineNativeEnum = class VineNativeEnum extends BaseLiteralType {
	static rules = { enum: enumRule };
	#values;
	[SUBTYPE] = "enum";
	constructor(values, options, validations) {
		super(options, validations || [enumRule({ choices: Object.values(values) })]);
		this.#values = values;
	}
	clone() {
		return new VineNativeEnum(this.#values, this.cloneOptions(), this.cloneValidations());
	}
};
const isNativeFileRule = createRule(function file(value, _, field) {
	if (!field.isDefined) return false;
	if (value instanceof File === false) {
		field.report(messages.nativeFile, "nativeFile", field);
		return false;
	}
	return true;
});
const minSizeRule = createRule(function minSize(value, options, field) {
	if (value.size < options.min) {
		field.report(messages["nativeFile.minSize"], "nativeFile.minSize", field, options);
		return false;
	}
	return true;
});
const maxSizeRule = createRule(function minSize(value, options, field) {
	if (value.size > options.max) field.report(messages["nativeFile.maxSize"], "nativeFile.maxSize", field, options);
});
const mimeTypesRule = createRule((value, options, field) => {
	const mimeType = value.type;
	if (!options.mimeTypes.includes(mimeType)) field.report(messages["nativeFile.mimeTypes"], "nativeFile.mimeTypes", field, options);
});
var VineNativeFile = class VineNativeFile extends BaseLiteralType {
	static rules = {
		maxSize: maxSizeRule,
		mimeTypes: mimeTypesRule,
		minSize: minSizeRule
	};
	[SUBTYPE] = "nativeFile";
	[UNIQUE_NAME] = "vine.nativeFile";
	[IS_OF_TYPE] = (value) => {
		return value instanceof File;
	};
	constructor(options, validations) {
		super(options, validations || []);
		this.dataTypeValidator = isNativeFileRule();
	}
	clone() {
		return new VineNativeFile(this.cloneOptions(), this.cloneValidations());
	}
	minSize(size) {
		return this.use(minSizeRule({ min: size }));
	}
	maxSize(size) {
		return this.use(maxSizeRule({ max: size }));
	}
	mimeTypes(types) {
		return this.use(mimeTypesRule({ mimeTypes: types }));
	}
};
var VineUnionOfTypes = class VineUnionOfTypes {
	#schemas;
	#otherwiseCallback = (_, field) => {
		field.report(messages.unionOfTypes, "unionOfTypes", field);
	};
	constructor(schemas) {
		this.#schemas = schemas;
	}
	otherwise(callback) {
		this.#otherwiseCallback = callback;
		return this;
	}
	clone() {
		const cloned = new VineUnionOfTypes(this.#schemas.map((schema) => schema.clone()));
		cloned.otherwise(this.#otherwiseCallback);
		return cloned;
	}
	optional() {
		return new VineUnionOfTypes([new VineOptional(), ...this.#schemas]);
	}
	nullable() {
		return new VineUnionOfTypes([new VineNull(), ...this.#schemas]);
	}
	toJSONSchema() {
		return { anyOf: this.#schemas.map((schema) => schema.toJSONSchema?.()).filter(Boolean) };
	}
	[PARSE](propertyName, refs, options) {
		return {
			type: "union",
			fieldName: propertyName,
			propertyName: options.toCamelCase ? camelcase(propertyName) : propertyName,
			elseConditionalFnRefId: refs.trackConditional(this.#otherwiseCallback),
			conditions: this.#schemas.map((schema) => {
				return {
					conditionalFnRefId: refs.trackConditional((value, field) => {
						return schema[IS_OF_TYPE](value, field);
					}),
					schema: schema[PARSE](propertyName, refs, options)
				};
			})
		};
	}
};
var SchemaBuilder = class extends Macroable {
	group = group;
	union = union;
	string() {
		return new VineString();
	}
	boolean(options) {
		return new VineBoolean(options);
	}
	accepted() {
		return new VineAccepted();
	}
	number(options) {
		return new VineNumber(options);
	}
	date(options) {
		return new VineDate(options);
	}
	literal(value) {
		return new VineLiteral(value);
	}
	optional() {
		return new VineOptional();
	}
	null() {
		return new VineNull();
	}
	object(properties) {
		return new VineObject(properties);
	}
	array(schema) {
		return new VineArray(schema);
	}
	tuple(schemas) {
		return new VineTuple(schemas);
	}
	record(schema) {
		return new VineRecord(schema);
	}
	enum(values) {
		if (Array.isArray(values) || typeof values === "function") return new VineEnum(values);
		return new VineNativeEnum(values);
	}
	any() {
		return new VineAny();
	}
	unionOfTypes(schemas) {
		const schemasInUse = /* @__PURE__ */ new Set();
		schemas.forEach((schema) => {
			if (!schema[IS_OF_TYPE] || !schema[UNIQUE_NAME]) throw new Error(`Cannot use "${schema.constructor.name}". The schema type is not compatible for use with "vine.unionOfTypes"`);
			if (schemasInUse.has(schema[UNIQUE_NAME])) throw new Error(`Cannot use duplicate schema "${schema[UNIQUE_NAME]}". "vine.unionOfTypes" needs distinct schema types only`);
			schemasInUse.add(schema[UNIQUE_NAME]);
		});
		schemasInUse.clear();
		return new VineUnionOfTypes(schemas);
	}
	nativeFile() {
		return new VineNativeFile();
	}
};
const COMPILER_ERROR_MESSAGES = {
	required: messages.required,
	array: messages.array,
	object: messages.object
};
const EMPTY_OBJECT = {};
var VineValidator = class {
	#compiled;
	#jsonSchema;
	"messagesProvider";
	"errorReporter";
	#parse(schema) {
		const refs = refsBuilder();
		return {
			compilerNode: {
				type: "root",
				schema: schema[PARSE]("", refs, { toCamelCase: false })
			},
			refs: refs.toJSON()
		};
	}
	"constructor"(schema, options) {
		this.schema = schema;
		this.options = options;
		const { compilerNode, refs } = this.#parse(schema);
		this.#compiled = {
			schema: compilerNode,
			refs
		};
		const metaDataValidator = options.metaDataValidator;
		const validateFn = new Compiler(compilerNode, {
			convertEmptyStringsToNull: options.convertEmptyStringsToNull,
			messages: COMPILER_ERROR_MESSAGES
		}).compile();
		this.errorReporter = options.errorReporter;
		this.messagesProvider = options.messagesProvider;
		if (metaDataValidator) this.validate = (data, validateOptions) => {
			let normalizedOptions = validateOptions ?? EMPTY_OBJECT;
			const meta = normalizedOptions.meta ?? {};
			const errorReporter = normalizedOptions.errorReporter ?? this.errorReporter;
			const messagesProvider = normalizedOptions.messagesProvider ?? this.messagesProvider;
			metaDataValidator(meta);
			return validateFn(data, meta, refs, messagesProvider, errorReporter());
		};
		else this.validate = (data, validateOptions) => {
			let normalizedOptions = validateOptions ?? EMPTY_OBJECT;
			const meta = normalizedOptions.meta ?? {};
			const errorReporter = normalizedOptions.errorReporter ?? this.errorReporter;
			return validateFn(data, meta, refs, normalizedOptions.messagesProvider ?? this.messagesProvider, errorReporter());
		};
	}
	async "tryValidate"(data, ...[options]) {
		try {
			return [null, await this.validate(data, options)];
		} catch (error) {
			if (error instanceof ValidationError) return [error, null];
			throw error;
		}
	}
	"toJSON"() {
		const { schema, refs } = this.#compiled;
		return {
			schema: structuredClone(schema),
			refs
		};
	}
	"toJSONSchema"() {
		if (!this.#jsonSchema) this.#jsonSchema = this.schema.toJSONSchema();
		return this.#jsonSchema;
	}
	"~standard" = {
		version: 1,
		vendor: "vinejs",
		jsonSchema: {
			input: () => {
				return this.toJSONSchema();
			},
			output: () => {
				throw new Error("Vine.js does not support creating validators using JSON Schema.");
			}
		},
		validate: async (data) => {
			const [error, result] = await this.tryValidate(data, {});
			if (result) return { value: result };
			return { issues: error?.messages.map((message) => {
				return {
					...message,
					path: message.field
				};
			}) };
		}
	};
};
var Vine = class extends SchemaBuilder {
	messagesProvider = new SimpleMessagesProvider(messages, fields);
	errorReporter = () => new SimpleErrorReporter();
	convertEmptyStringsToNull = false;
	helpers = helpers;
	createRule = createRule;
	compile(schema) {
		return new VineValidator(schema, {
			convertEmptyStringsToNull: this.convertEmptyStringsToNull,
			messagesProvider: this.messagesProvider,
			errorReporter: this.errorReporter
		});
	}
	create(schemaOrProperties) {
		const validatorOptions = {
			convertEmptyStringsToNull: this.convertEmptyStringsToNull,
			messagesProvider: this.messagesProvider,
			errorReporter: this.errorReporter
		};
		if (PARSE in schemaOrProperties) return new VineValidator(schemaOrProperties, validatorOptions);
		return new VineValidator(new VineObject(schemaOrProperties), validatorOptions);
	}
	withMetaData(callback) {
		return {
			compile: (schema) => {
				return new VineValidator(schema, {
					convertEmptyStringsToNull: this.convertEmptyStringsToNull,
					messagesProvider: this.messagesProvider,
					errorReporter: this.errorReporter,
					metaDataValidator: callback
				});
			},
			create: (schemaOrProperties) => {
				const validatorOptions = {
					convertEmptyStringsToNull: this.convertEmptyStringsToNull,
					messagesProvider: this.messagesProvider,
					errorReporter: this.errorReporter,
					metaDataValidator: callback
				};
				if (PARSE in schemaOrProperties) return new VineValidator(schemaOrProperties, validatorOptions);
				return new VineValidator(new VineObject(schemaOrProperties), validatorOptions);
			}
		};
	}
	validate(options) {
		return this.compile(options.schema).validate(options.data, options);
	}
	tryValidate(options) {
		return this.compile(options.schema).tryValidate(options.data, options);
	}
};
const vine = new Vine();
export { BaseLiteralType, BaseType, ConditionalValidations, MetaModifier, NullableModifier, OptionalModifier, SimpleErrorReporter, SimpleMessagesProvider, UnionConditional, ValidationError, Vine, VineAccepted, VineAny, VineArray, VineBoolean, VineDate, VineEnum, VineLiteral, VineNativeEnum, VineNativeFile, VineNull, VineNumber, VineObject, VineOptional, VineRecord, VineString, VineTuple, VineUnion, VineValidator, vine as default, main_exports as errors, requiredWhen, symbols_exports as symbols };
