// @vinejs/vine/build/simple_error_reporter-DjITOpes.js
import "node:module";
import delve from "dlv";
import dayjs from "dayjs";
import isIP from "validator/lib/isIP.js";
import isJWT from "validator/lib/isJWT.js";
import isURL from "validator/lib/isURL.js";
import isSlug from "validator/lib/isSlug.js";
import isIBAN from "validator/lib/isIBAN.js";
import isUUID from "validator/lib/isUUID.js";
import isAscii from "validator/lib/isAscii.js";
import isEmail from "validator/lib/isEmail.js";
import isAlpha from "validator/lib/isAlpha.js";
import isLatLong from "validator/lib/isLatLong.js";
import isDecimal from "validator/lib/isDecimal.js";
import isHexColor from "validator/lib/isHexColor.js";
import isCreditCard from "validator/lib/isCreditCard.js";
import isSameOrAfter from "dayjs/plugin/isSameOrAfter.js";
import isSameOrBefore from "dayjs/plugin/isSameOrBefore.js";
import isAlphanumeric from "validator/lib/isAlphanumeric.js";
import isPassportNumber from "validator/lib/isPassportNumber.js";
import isVAT from "validator/lib/isVAT.js";
import customParseFormat from "dayjs/plugin/customParseFormat.js";
import isPostalCode, { locales } from "validator/lib/isPostalCode.js";
import isMobilePhone, { locales as locales$1 } from "validator/lib/isMobilePhone.js";
var __defProp = Object.defineProperty;
var __exportAll = (all, no_symbols) => {
	let target = {};
	for (var name in all) __defProp(target, name, {
		get: all[name],
		enumerable: true
	});
	if (!no_symbols) __defProp(target, Symbol.toStringTag, { value: "Module" });
	return target;
};
const BOOLEAN_POSITIVES = [
	"1",
	1,
	"true",
	true,
	"on"
];
const BOOLEAN_NEGATIVES = [
	"0",
	0,
	"false",
	false
];
const DEFAULT_DATE_FORMATS = ["YYYY-MM-DD", "YYYY-MM-DD HH:mm:ss"];
const ULID = /^[0-9A-HJKMNP-TV-Za-hjkmnp-tv-z]{26}$/;
dayjs.extend(customParseFormat);
dayjs.extend(isSameOrAfter);
dayjs.extend(isSameOrBefore);
const helpers = {
	exists(value) {
		return value !== null && value !== void 0;
	},
	isMissing(value) {
		return !this.exists(value);
	},
	isTrue(value) {
		return BOOLEAN_POSITIVES.includes(value);
	},
	isFalse(value) {
		return BOOLEAN_NEGATIVES.includes(value);
	},
	isString(value) {
		return typeof value === "string";
	},
	isObject(value) {
		return !!(value && typeof value === "object" && !Array.isArray(value));
	},
	hasKeys(value, keys) {
		for (let key of keys) if (key in value === false) return false;
		return true;
	},
	isArray(value) {
		return Array.isArray(value);
	},
	isNumeric(value) {
		return !Number.isNaN(Number(value));
	},
	asNumber(value) {
		return value === null ? NaN : Number(value);
	},
	asBoolean(value) {
		if (this.isTrue(value)) return true;
		if (this.isFalse(value)) return false;
		return null;
	},
	asDayJS(value, format) {
		let isTimestampAllowed = false;
		let isISOAllowed = false;
		let formats = format || DEFAULT_DATE_FORMATS;
		if (Array.isArray(formats)) {
			formats = [...formats];
			isTimestampAllowed = formats.includes("x");
			isISOAllowed = formats.includes("iso8601");
		} else if (typeof formats !== "string") {
			formats = { ...formats };
			isTimestampAllowed = formats.format === "x";
			isISOAllowed = formats.format === "iso";
		}
		const valueAsNumber = isTimestampAllowed ? helpers.asNumber(value) : value;
		let dateTime;
		if (isTimestampAllowed && !Number.isNaN(valueAsNumber)) dateTime = dayjs(valueAsNumber);
		else dateTime = dayjs(value, formats, true);
		if (!dateTime.isValid() && isISOAllowed) dateTime = dayjs(value);
		return {
			dateTime,
			formats
		};
	},
	compareValues(inputValue, expectedValue) {
		let input = inputValue;
		if (typeof expectedValue === "boolean") input = this.asBoolean(inputValue);
		else if (typeof expectedValue === "number") input = this.asNumber(inputValue);
		return {
			isEqual: input === expectedValue,
			casted: input
		};
	},
	isEmail: isEmail.default,
	isURL: isURL.default,
	isAlpha: isAlpha.default,
	isAlphaNumeric: isAlphanumeric.default,
	isIP: isIP.default,
	isUUID: isUUID.default,
	isAscii: isAscii.default,
	isCreditCard: isCreditCard.default,
	isIBAN: isIBAN.default,
	isJWT: isJWT.default,
	isLatLong: isLatLong.default,
	isMobilePhone: isMobilePhone.default,
	isPassportNumber: isPassportNumber.default,
	isVAT: isVAT.default,
	isPostalCode: isPostalCode.default,
	isSlug: isSlug.default,
	isDecimal: isDecimal.default,
	mobileLocales: locales$1,
	postalCountryCodes: locales,
	passportCountryCodes: [
		"AM",
		"AR",
		"AT",
		"AU",
		"AZ",
		"BE",
		"BG",
		"BR",
		"BY",
		"CA",
		"CH",
		"CY",
		"CZ",
		"DE",
		"DK",
		"DZ",
		"ES",
		"FI",
		"FR",
		"GB",
		"GR",
		"HR",
		"HU",
		"IE",
		"IN",
		"ID",
		"IR",
		"IS",
		"IT",
		"JM",
		"JP",
		"KR",
		"KZ",
		"LI",
		"LT",
		"LU",
		"LV",
		"LY",
		"MT",
		"MZ",
		"MY",
		"MX",
		"NL",
		"NZ",
		"PH",
		"PK",
		"PL",
		"PT",
		"RO",
		"RU",
		"SE",
		"SL",
		"SK",
		"TH",
		"TR",
		"UA",
		"US"
	],
	isULID(value) {
		if (typeof value !== "string") return false;
		if (value[0] > "7") return false;
		return ULID.test(value);
	},
	isHexColor: (value) => {
		if (!value.startsWith("#")) return false;
		return isHexColor.default(value);
	},
	isActiveURL: async (url) => {
		const { resolve4, resolve6 } = await import("node:dns/promises");
		try {
			const { hostname } = new URL(url);
			if ((await resolve6(hostname)).length) return true;
			else return (await resolve4(hostname)).length > 0;
		} catch {
			return false;
		}
	},
	isDistinct: (dataSet, fields) => {
		const uniqueItems = /* @__PURE__ */ new Set();
		if (!fields) {
			for (let item of dataSet) if (helpers.exists(item)) if (uniqueItems.has(item)) return false;
			else uniqueItems.add(item);
			return true;
		}
		const fieldsList = Array.isArray(fields) ? fields : [fields];
		for (let item of dataSet) if (helpers.isObject(item) && helpers.hasKeys(item, fieldsList)) {
			const element = fieldsList.map((field) => item[field]).join("_");
			if (uniqueItems.has(element)) return false;
			else uniqueItems.add(element);
		}
		return true;
	},
	getNestedValue(key, field) {
		if (key.indexOf(".") > -1) return delve(field.data, key);
		return field.parent[key];
	},
	optional(props) {
		const result = {};
		for (const name of Object.keys(props)) {
			let field = props[name].clone();
			if ("optional" in field && typeof field.optional === "function") field = field.optional();
			result[name] = field;
		}
		return result;
	}
};
var SimpleMessagesProvider = class {
	#messages;
	#fields;
	constructor(messages, fields) {
		this.#messages = messages;
		this.#fields = fields || {};
	}
	#interpolate(message, data) {
		if (!message.includes("{{")) return message;
		return message.replace(/(\\)?{{(.*?)}}/g, (_, __, key) => {
			const tokens = key.trim().split(".");
			let output = data;
			while (tokens.length) {
				if (output === null || typeof output !== "object") return;
				const token = tokens.shift();
				output = Object.hasOwn(output, token) ? output[token] : void 0;
			}
			return output;
		});
	}
	getMessage(rawMessage, rule, field, args) {
		const fieldName = this.#fields[field.name] || field.name;
		const fieldMessage = this.#messages[`${field.getFieldPath()}.${rule}`];
		if (fieldMessage) return this.#interpolate(fieldMessage, {
			field: fieldName,
			...args
		});
		const wildcardMessage = this.#messages[`${field.wildCardPath}.${rule}`];
		if (wildcardMessage) return this.#interpolate(wildcardMessage, {
			field: fieldName,
			...args
		});
		const ruleMessage = this.#messages[rule];
		if (ruleMessage) return this.#interpolate(ruleMessage, {
			field: fieldName,
			...args
		});
		return this.#interpolate(rawMessage, {
			field: fieldName,
			...args
		});
	}
	toJSON() {
		return {
			messages: this.#messages,
			fields: this.#fields
		};
	}
};
var ValidationError = class extends Error {
	status = 422;
	code = "E_VALIDATION_ERROR";
	constructor(messages, options) {
		super("Validation failure", options);
		this.messages = messages;
		const ErrorConstructor = this.constructor;
		if ("captureStackTrace" in Error) Error.captureStackTrace(this, ErrorConstructor);
	}
	get [Symbol.toStringTag]() {
		return this.constructor.name;
	}
	toString() {
		return `${this.name} [${this.code}]: ${this.message}`;
	}
};
var main_exports = /* @__PURE__ */ __exportAll({ E_VALIDATION_ERROR: () => E_VALIDATION_ERROR });
const E_VALIDATION_ERROR = ValidationError;
var SimpleErrorReporter = class {
	hasErrors = false;
	errors = [];
	report(message, rule, field, meta) {
		const error = {
			message,
			rule,
			field: field.getFieldPath()
		};
		if (meta) error.meta = meta;
		if (field.isArrayMember) error.index = field.name;
		this.hasErrors = true;
		this.errors.push(error);
	}
	createError() {
		return new E_VALIDATION_ERROR(this.errors);
	}
};
export { BOOLEAN_NEGATIVES as a, __exportAll as c, SimpleMessagesProvider as i, main_exports as n, BOOLEAN_POSITIVES as o, ValidationError as r, helpers as s, SimpleErrorReporter as t };
