// @japa/plugin-adonisjs/build/extend_api_client-YWKOU40C.js
import { t as debug_default } from "./debug-cDi3H0Pi.js";
import { ApiClient, ApiRequest } from "@japa/api-client";
import { createURL } from "@adonisjs/core/helpers/http";
function extendApiClient(cookieClient, router) {
	debug_default("extending @japa/api-client with adonisjs specific methods");
	ApiClient.setRouteBuilder((name, params) => {
		const route = router.findOrFail(name);
		return {
			url: createURL(route.pattern, route.tokens, router.qs.stringify, params),
			method: route.methods[0]
		};
	});
	ApiClient.cookiesSerializer({
		prepare(_, value) {
			return value;
		},
		process(key, value) {
			if (!value) return value;
			return cookieClient.parse(key, value);
		}
	});
	ApiRequest.macro("withCookie", function(key, value) {
		const signedValue = cookieClient.sign(key, value);
		if (signedValue) this.cookiesJar[key] = {
			name: key,
			value: signedValue
		};
		return this;
	});
	ApiRequest.macro("cookie", function(key, value) {
		return this.withCookie(key, value);
	});
	ApiRequest.macro("withEncryptedCookie", function(key, value) {
		const encryptedValue = cookieClient.encrypt(key, value);
		if (encryptedValue) this.cookiesJar[key] = {
			name: key,
			value: encryptedValue
		};
		return this;
	});
	ApiRequest.macro("encryptedCookie", function(key, value) {
		return this.withEncryptedCookie(key, value);
	});
	ApiRequest.macro("withPlainCookie", function(key, value) {
		const plainValue = cookieClient.encode(key, value);
		if (plainValue) this.cookiesJar[key] = {
			name: key,
			value: plainValue
		};
		return this;
	});
	ApiRequest.macro("plainCookie", function(key, value) {
		return this.withPlainCookie(key, value);
	});
}
export { extendApiClient };
