// @adonisjs/vite/build/providers/vite_provider.js
import {
  Vite
} from "../chunk-ZCIGOANY.js";
import {
  ViteMiddleware
} from "../chunk-AF6PV64J.js";
import "../chunk-ZCQTQOMI.js";

// providers/vite_provider.ts
var ViteProvider = class {
  /**
   * Creates a new ViteProvider instance
   *
   * @param app - The AdonisJS application service instance
   *
   * @example
   * const provider = new ViteProvider(app)
   */
  constructor(app) {
    this.app = app;
  }
  /**
   * Flag indicating whether the Vite development server should be started
   * Set to true when manifest file doesn't exist and running in web/test environment
   */
  #shouldRunViteDevServer = false;
  /**
   * Registers the Vite Edge.js plugin if Edge.js is available in the application
   *
   * Adds global helpers and custom tags (@vite, @viteReactRefresh) to Edge templates.
   * Only registers the plugin if the application is using Edge.js.
   *
   * @example
   * await provider.registerEdgePlugin()
   * // Enables @vite('app.js') in Edge templates
   */
  async registerEdgePlugin() {
    if (this.app.usingEdgeJS) {
      const edge = await import("edge.js");
      const vite = await this.app.container.make("vite");
      const { edgePluginVite } = await import("../src/plugins/edge.js");
      edge.default.use(edgePluginVite(vite));
    }
  }
  /**
   * Registers Content Security Policy keywords when @adonisjs/shield is installed
   *
   * Adds the '@viteUrl' keyword for CSP directives that returns the Vite assets URL
   * when it's an HTTP/HTTPS URL, or an empty string otherwise.
   *
   * @example
   * await provider.registerShieldKeywords()
   * // Enables @viteUrl in CSP directives
   */
  async registerShieldKeywords() {
    let cspKeywords = null;
    try {
      const shieldExports = await import("@adonisjs/shield");
      cspKeywords = shieldExports.cspKeywords;
    } catch {
    }
    if (!cspKeywords) {
      return;
    }
    const vite = await this.app.container.make("vite");
    cspKeywords.register("@viteUrl", function() {
      const assetsURL = vite.assetsUrl();
      if (!assetsURL || !assetsURL.startsWith("http://") || assetsURL.startsWith("https://")) {
        return "";
      }
      return assetsURL;
    });
  }
  /**
   * Registers Vite service and middleware bindings in the IoC container
   *
   * Creates a Vite instance using configuration and determines whether
   * the development server should run based on environment and manifest file presence.
   *
   * @example
   * provider.register()
   * // Vite service is now available via container.make('vite')
   */
  register() {
    const appEnvironment = this.app.getEnvironment();
    const vite = new Vite(this.app.config.get("vite"));
    this.#shouldRunViteDevServer = appEnvironment === "test" || !!process.env.DEV_MODE;
    this.app.container.singleton(Vite, () => vite);
    this.app.container.singleton(ViteMiddleware, () => new ViteMiddleware(vite));
    this.app.container.alias("vite", Vite);
  }
  /**
   * Boots the Vite provider by registering plugins and integrations
   *
   * Registers Edge.js plugin and Shield CSP keywords if the respective
   * packages are available in the application.
   *
   * @example
   * await provider.boot()
   */
  async boot() {
    await this.registerEdgePlugin();
    await this.registerShieldKeywords();
  }
  /**
   * Starts the Vite development server when the application is ready
   *
   * Only starts the server if running in development/test mode and
   * no manifest file exists (indicating development mode).
   *
   * @example
   * await provider.ready()
   * // Dev server starts on configured port
   */
  async ready() {
    if (!this.#shouldRunViteDevServer) {
      return;
    }
    const vite = await this.app.container.make("vite");
    await vite.createDevServer();
  }
  /**
   * Gracefully stops the Vite development server during application shutdown
   *
   * Only attempts to stop the server if it was started during the ready phase.
   * Ensures clean shutdown of the development server and its resources.
   *
   * @example
   * await provider.shutdown()
   */
  async shutdown() {
    if (!this.#shouldRunViteDevServer) {
      return;
    }
    const vite = await this.app.container.make("vite");
    await vite.stopDevServer();
  }
};
export {
  ViteProvider as default
};
