// @adonisjs/assembler/build/src/helpers.js
import { parseImports } from "parse-imports";
async function findImport(code, importReference) {
	const importIdentifier = importReference.split(".")[0];
	for (const $import of await parseImports(code, {})) {
		if (!$import.importClause) continue;
		if (!$import.moduleSpecifier.value) continue;
		if ($import.importClause.default === importIdentifier) return {
			specifier: $import.moduleSpecifier.value,
			isConstant: $import.moduleSpecifier.isConstant,
			clause: {
				type: "default",
				value: importIdentifier
			}
		};
		if ($import.importClause.namespace === importIdentifier) return {
			specifier: $import.moduleSpecifier.value,
			isConstant: $import.moduleSpecifier.isConstant,
			clause: {
				type: "namespace",
				value: importIdentifier
			}
		};
		const namedImport = $import.importClause.named.find(({ binding }) => {
			return binding === importIdentifier;
		});
		if (namedImport) return {
			specifier: $import.moduleSpecifier.value,
			isConstant: $import.moduleSpecifier.isConstant,
			clause: {
				type: "named",
				value: namedImport.specifier,
				...namedImport.binding !== namedImport.specifier ? { alias: namedImport.binding } : {}
			}
		};
	}
	return null;
}
function inspectClass(node) {
	return node.find({ rule: { kind: "class_declaration" } });
}
function inspectClassMethods(node) {
	return node.findAll({ rule: { kind: "method_definition" } });
}
function nodeToPlainText(node) {
	let out = [];
	function toText(one) {
		const children = one.children();
		if (!children.length) out.push(one.text());
		else children.forEach((child) => toText(child));
	}
	toText(node);
	return out.join("");
}
function inspectMethodArguments(node, methodCalls) {
	return node.findAll({ rule: { any: methodCalls.map((methodCall) => {
		return { pattern: {
			context: `${methodCall}($$$ARGUMENTS)`,
			selector: "call_expression"
		} };
	}) } }).flatMap((matchingExpression) => {
		return matchingExpression.findAll({ rule: { kind: "arguments" } });
	});
}
function searchValidatorDirectUsage(node) {
	return node.findAll({ rule: { pattern: {
		context: "$$$VALIDATOR.validate($$$)",
		selector: "call_expression"
	} } }).flatMap((expression) => {
		return expression.getMultipleMatches("VALIDATOR");
	});
}
export { findImport, inspectClass, inspectClassMethods, inspectMethodArguments, nodeToPlainText, searchValidatorDirectUsage };
