// @adonisjs/assembler/build/codemod_exception-CzQgXAAf.js
var CodemodException = class CodemodException extends Error {
	instructions;
	filePath;
	constructor(message, options) {
		super(message);
		this.name = "CodemodException";
		this.instructions = options?.instructions;
		this.filePath = options?.filePath;
	}
	static #formatEnvValidations(definition) {
		const lines = [];
		if (definition.leadingComment) {
			lines.push(`/*`);
			lines.push(`|----------------------------------------------------------`);
			lines.push(`| ${definition.leadingComment}`);
			lines.push(`|----------------------------------------------------------`);
			lines.push(`*/`);
		}
		for (const [variable, validation] of Object.entries(definition.variables)) lines.push(`${variable}: ${validation},`);
		return lines.join("\n");
	}
	static #formatMiddleware(stack, middleware) {
		if (stack === "named") return `export const middleware = router.named({\n  ${middleware.filter((m) => m.name).map((m) => `${m.name}: () => import('${m.path}')`).join(",\n  ")}\n})`;
		return `${stack}.use([\n  ${middleware.map((m) => `() => import('${m.path}')`).join(",\n  ")}\n])`;
	}
	static #formatPolicies(policies) {
		return `export const policies = {\n  ${policies.map((p) => `${p.name}: () => import('${p.path}')`).join(",\n  ")}\n}`;
	}
	static #formatVitePlugin(pluginCall, importDeclarations) {
		return `${importDeclarations.map((decl) => decl.isNamed ? `import { ${decl.identifier} } from '${decl.module}'` : `import ${decl.identifier} from '${decl.module}'`).join("\n")}\n\nexport default defineConfig({\n  plugins: [${pluginCall}]\n})`;
	}
	static #formatJapaPlugin(pluginCall, importDeclarations) {
		return `${importDeclarations.map((decl) => decl.isNamed ? `import { ${decl.identifier} } from '${decl.module}'` : `import ${decl.identifier} from '${decl.module}'`).join("\n")}\n\nexport const plugins: Config['plugins'] = [\n  ${pluginCall}\n]`;
	}
	static missingEnvFile(filePath, definition) {
		const code = this.#formatEnvValidations(definition);
		return new CodemodException(`Could not find source file at path: "${filePath}"`, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${code}`
		});
	}
	static missingEnvCreate(filePath, definition) {
		return new CodemodException(`Cannot find Env.create statement in the file.`, {
			filePath,
			instructions: `Add the following code inside Env.create() in "${filePath}":\n\n${this.#formatEnvValidations(definition)}`
		});
	}
	static invalidEnvCreate(filePath, definition) {
		return new CodemodException(`The second argument of Env.create is not an object literal.`, {
			filePath,
			instructions: `Add the following code inside Env.create() in "${filePath}":\n\n${this.#formatEnvValidations(definition)}`
		});
	}
	static missingKernelFile(filePath, stack, middleware) {
		const code = this.#formatMiddleware(stack, middleware);
		return new CodemodException(`Could not find source file at path: "${filePath}"`, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${code}`
		});
	}
	static missingMiddlewareStack(filePath, stack, middleware) {
		const code = this.#formatMiddleware(stack, middleware);
		return new CodemodException(`Cannot find ${stack === "named" ? "middleware variable" : `${stack}.use`} statement in the file.`, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${code}`
		});
	}
	static invalidMiddlewareStack(filePath, stack, middleware, reason) {
		return new CodemodException(reason, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${this.#formatMiddleware(stack, middleware)}`
		});
	}
	static missingPoliciesFile(filePath, policies) {
		const code = this.#formatPolicies(policies);
		return new CodemodException(`Could not find source file at path: "${filePath}"`, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${code}`
		});
	}
	static invalidPoliciesFile(filePath, policies, reason) {
		return new CodemodException(reason, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${this.#formatPolicies(policies)}`
		});
	}
	static missingViteConfig(filePath, pluginCall, importDeclarations) {
		return new CodemodException(`Cannot find vite.config.ts file. Make sure to rename vite.config.js to vite.config.ts`, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${this.#formatVitePlugin(pluginCall, importDeclarations)}`
		});
	}
	static invalidViteConfig(filePath, pluginCall, importDeclarations, reason) {
		return new CodemodException(reason, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${this.#formatVitePlugin(pluginCall, importDeclarations)}`
		});
	}
	static missingJapaBootstrap(filePath, pluginCall, importDeclarations) {
		const code = this.#formatJapaPlugin(pluginCall, importDeclarations);
		return new CodemodException(`Could not find source file at path: "${filePath}"`, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${code}`
		});
	}
	static invalidJapaBootstrap(filePath, pluginCall, importDeclarations, reason) {
		return new CodemodException(reason, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${this.#formatJapaPlugin(pluginCall, importDeclarations)}`
		});
	}
	static missingRcFile(filePath, codeToAdd) {
		return new CodemodException(`Could not find source file at path: "${filePath}"`, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${codeToAdd}`
		});
	}
	static invalidRcFile(filePath, codeToAdd, reason) {
		return new CodemodException(reason, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${codeToAdd}`
		});
	}
	static E_CODEMOD_FILE_NOT_FOUND(filePath, codeToAdd) {
		return new CodemodException(`Could not find source file at path: "${filePath}"`, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${codeToAdd}`
		});
	}
	static E_CODEMOD_INVALID_STRUCTURE(message, filePath, codeToAdd) {
		return new CodemodException(message, {
			filePath,
			instructions: `Add the following code to "${filePath}":\n\n${codeToAdd}`
		});
	}
};
export { CodemodException as t };
