// @adonisjs/assembler/build/index.js
import { a as loadHooks, c as readTsConfig, d as runNode, f as throttle, m as debug_default, n as copyFiles, o as memoize, p as watch, r as getPort, s as parseConfig, t as VirtualFileSystem, u as run } from "./virtual_file_system-bGeoWsK-.js";
import { n as FileBuffer } from "./source-dVeugJ0e.js";
import { IndexGenerator } from "./src/index_generator/main.js";
import "./validator_extractor-Ccio_Ndi.js";
import { RoutesScanner } from "./src/code_scanners/routes_scanner/main.js";
import { t as CodemodException } from "./codemod_exception-CzQgXAAf.js";
import dedent from "dedent";
import fs, { readFile, unlink } from "node:fs/promises";
import { cliui } from "@poppinss/cliui";
import { fileURLToPath } from "node:url";
import string from "@poppinss/utils/string";
import { join, relative } from "node:path/posix";
import { detectPackageManager } from "@antfu/install-pkg";
import getRandomPort from "get-port";
import picomatch from "picomatch";
import prettyHrtime from "pretty-hrtime";
import { RuntimeException } from "@poppinss/utils/exception";
const SUPPORTED_PACKAGE_MANAGERS = {
	"npm": {
		packageManagerFiles: ["package-lock.json"],
		installCommand: "npm ci --omit=\"dev\""
	},
	"yarn": {
		packageManagerFiles: ["yarn.lock"],
		installCommand: "yarn install --production"
	},
	"yarn@berry": {
		packageManagerFiles: [
			"yarn.lock",
			".yarn/**/*",
			".yarnrc.yml"
		],
		installCommand: "yarn workspaces focus --production"
	},
	"pnpm": {
		packageManagerFiles: ["pnpm-lock.yaml"],
		installCommand: "pnpm i --prod"
	},
	"bun": {
		packageManagerFiles: ["bun.lockb"],
		installCommand: "bun install --production"
	}
};
var Bundler = class {
	#ts;
	#hooks;
	#indexGenerator;
	ui = cliui();
	cwd;
	cwdPath;
	options;
	constructor(cwd, ts, options) {
		this.cwd = cwd;
		this.options = options;
		this.cwdPath = string.toUnixSlash(fileURLToPath(this.cwd));
		this.#ts = ts;
	}
	#getRelativeName(filePath) {
		return string.toUnixSlash(relative(this.cwdPath, filePath));
	}
	async #cleanupBuildDirectory(outDir) {
		await fs.rm(outDir, {
			recursive: true,
			force: true,
			maxRetries: 5
		});
	}
	async #runTsc(outDir) {
		try {
			await run(this.cwd, {
				stdio: "inherit",
				script: "tsc",
				scriptArgs: ["--outDir", outDir]
			});
			return true;
		} catch {
			return false;
		}
	}
	async #copyMetaFiles(outDir, additionalFilesToCopy) {
		await copyFiles((this.options.metaFiles || []).map((file) => file.pattern).concat(additionalFilesToCopy), this.cwdPath, outDir);
	}
	async #detectPackageManager() {
		const pkgManager = await detectPackageManager(this.cwdPath);
		if (pkgManager === "deno") return "npm";
		if (pkgManager === "pnpm@6") return "pnpm";
		return pkgManager;
	}
	async #createAceFile(outDir) {
		const aceFileLocation = join(outDir, "ace.js");
		const aceFileContent = dedent(`
      /**
       * This file is auto-generated by the build process.
       * If you had any custom code inside this file, then
       * instead write it inside the "bin/console.js" file.
       */

      await import('./bin/console.js')
    `);
		await fs.writeFile(aceFileLocation, aceFileContent);
		this.ui.logger.info("created ace file", { suffix: this.#getRelativeName(aceFileLocation) });
	}
	async bundle(stopOnError = true, client) {
		this.packageManager = client ?? await this.#detectPackageManager() ?? "npm";
		const config = parseConfig(this.cwd, this.#ts);
		if (!config) return false;
		this.ui.logger.info("loading hooks...");
		this.#hooks = await loadHooks(this.options.hooks, [
			"init",
			"buildStarting",
			"buildFinished"
		]);
		this.#indexGenerator = new IndexGenerator(this.cwdPath, this.ui.logger);
		await this.#hooks.runner("init").run(this, this.#hooks, this.#indexGenerator);
		this.#hooks.clear("init");
		this.ui.logger.info("generating indexes...");
		await this.#indexGenerator.generate();
		const outDir = config.options.outDir || fileURLToPath(new URL("build/", this.cwd));
		this.ui.logger.info("cleaning up output directory", { suffix: this.#getRelativeName(outDir) });
		await this.#cleanupBuildDirectory(outDir);
		await this.#hooks.runner("buildStarting").run(this);
		this.ui.logger.info("compiling typescript source", { suffix: "tsc" });
		const buildCompleted = await this.#runTsc(outDir);
		await this.#createAceFile(outDir);
		if (!buildCompleted && stopOnError) {
			await this.#cleanupBuildDirectory(outDir);
			const instructions = this.ui.sticker().fullScreen().drawBorder((borderChar, colors) => colors.red(borderChar));
			instructions.add(this.ui.colors.red("Cannot complete the build process as there are TypeScript errors."));
			instructions.add(this.ui.colors.red("Use \"--ignore-ts-errors\" flag to ignore TypeScript errors and continue the build."));
			this.ui.logger.logError(instructions.prepare());
			return false;
		}
		const pkgFiles = ["package.json", ...SUPPORTED_PACKAGE_MANAGERS[this.packageManager].packageManagerFiles];
		this.ui.logger.info("copying meta files to the output directory");
		await this.#copyMetaFiles(outDir, pkgFiles);
		this.ui.logger.success("build completed");
		this.ui.logger.log("");
		const displayMessage = this.ui.instructions().heading("Run the following commands to start the server in production");
		await this.#hooks.runner("buildFinished").run(this, displayMessage);
		displayMessage.add(this.ui.colors.cyan(`cd ${this.#getRelativeName(outDir)}`)).add(this.ui.colors.cyan(SUPPORTED_PACKAGE_MANAGERS[this.packageManager].installCommand)).add(this.ui.colors.cyan("node bin/server.js")).render();
		return true;
	}
};
const DEFAULT_INCLUDES = ["**/*"];
const ALWAYS_EXCLUDE = [
	".git/**",
	"coverage/**",
	".github/**",
	".adonisjs/**"
];
const DEFAULT_EXCLUDES = [
	"node_modules/**",
	"bower_components/**",
	"jspm_packages/**"
];
var FileSystem = class {
	#cwd;
	#tsConfig;
	#scannedTypeScriptFiles = /* @__PURE__ */ new Set();
	#isIncluded;
	#isExcluded;
	#isMetaFileWithReloadsEnabled;
	#isMetaFileWithReloadsDisabled;
	#isTestFile;
	#includes;
	#excludes;
	get includes() {
		return this.#includes;
	}
	get excludes() {
		return this.#excludes;
	}
	inspect = memoize((absolutePath, relativePath) => {
		relativePath = relativePath ?? relative(this.#cwd, absolutePath);
		if (this.#isScriptFile(relativePath) && (this.#scannedTypeScriptFiles.has(absolutePath) || this.#isPartOfBackendProject(relativePath))) {
			debug_default("backend project file \"%s\"", relativePath);
			const isTestFile = this.#isTestFile(relativePath);
			return {
				fileType: isTestFile ? "test" : "script",
				reloadServer: !isTestFile,
				unixRelativePath: relativePath,
				unixAbsolutePath: absolutePath
			};
		}
		if (this.#isMetaFileWithReloadsEnabled(relativePath)) {
			debug_default("meta file \"%s\"", relativePath);
			return {
				fileType: "meta",
				reloadServer: true,
				unixRelativePath: relativePath,
				unixAbsolutePath: absolutePath
			};
		}
		if (this.#isMetaFileWithReloadsDisabled(relativePath)) {
			debug_default("meta file \"%s\"", relativePath);
			return {
				fileType: "meta",
				reloadServer: false,
				unixRelativePath: relativePath,
				unixAbsolutePath: absolutePath
			};
		}
		debug_default("ignored file \"%s\"", relativePath);
		return null;
	});
	shouldWatchDirectory = memoize((absolutePath) => {
		if (absolutePath === this.#cwd) {
			debug_default("watching project root");
			return true;
		}
		const relativePath = relative(this.#cwd, absolutePath);
		if (this.#isExcluded(relativePath)) {
			debug_default("watching \"%s\"", absolutePath);
			return false;
		}
		return true;
	});
	constructor(cwd, tsConfig, rcFile) {
		this.#cwd = cwd;
		this.#tsConfig = tsConfig;
		const files = tsConfig.config.files ?? [];
		const metaFiles = rcFile.metaFiles ?? [];
		const testSuites = rcFile.suites ?? [];
		const outDir = tsConfig.config.compilerOptions?.outDir;
		for (const file of files) this.#scannedTypeScriptFiles.add(string.toUnixSlash(file));
		this.#includes = tsConfig.config.include || DEFAULT_INCLUDES;
		this.#excludes = ALWAYS_EXCLUDE.concat(tsConfig.config.exclude || (outDir ? DEFAULT_EXCLUDES.concat(outDir) : DEFAULT_EXCLUDES));
		const metaFilesWithReloads = [];
		const metaFilesWithoutReloads = [];
		for (const file of metaFiles) if (file.reloadServer) metaFilesWithReloads.push(file.pattern);
		else metaFilesWithoutReloads.push(file.pattern);
		const testFilePatterns = testSuites.flatMap((suite) => suite.files);
		const picomatcchOptions = { cwd: this.#cwd };
		this.#isMetaFileWithReloadsEnabled = picomatch(metaFilesWithReloads, picomatcchOptions);
		this.#isMetaFileWithReloadsDisabled = picomatch(metaFilesWithoutReloads, picomatcchOptions);
		this.#isTestFile = picomatch(testFilePatterns, picomatcchOptions);
		this.#isIncluded = picomatch(this.#includes, picomatcchOptions);
		this.#isExcluded = picomatch(this.#excludes, picomatcchOptions);
		debug_default("initiating file system %O", {
			includes: this.#includes,
			excludes: this.#excludes,
			outDir,
			files,
			metaFiles,
			testSuites
		});
	}
	#isScriptFile(relativePath) {
		if ((relativePath.endsWith(".ts") || relativePath.endsWith(".tsx")) && !relativePath.endsWith(".d.ts")) return true;
		if (this.#tsConfig.config.compilerOptions?.allowJs && relativePath.endsWith(".js")) return true;
		if (this.#tsConfig.config.compilerOptions?.resolveJsonModule && relativePath.endsWith(".json")) return true;
		return false;
	}
	#isPartOfBackendProject(relativePath) {
		if (this.#isExcluded(relativePath)) {
			debug_default("excluded by tsconfig \"%s\"", relativePath);
			return false;
		}
		if (this.#isIncluded(relativePath)) {
			debug_default("included by tsconfig \"%s\"", relativePath);
			return true;
		}
		return false;
	}
	shouldWatchFile(absolutePath) {
		return this.inspect(absolutePath) !== null;
	}
};
var ShortcutsManager = class {
	#logger;
	#callbacks;
	#serverUrl;
	#keyPressHandler;
	#shortcuts = [
		{
			key: "r",
			description: "restart server",
			handler: () => {
				this.#logger.log("");
				this.#logger.info("Manual restart triggered...");
				this.#callbacks.onRestart();
			}
		},
		{
			key: "c",
			description: "clear console",
			handler: () => {
				this.#callbacks.onClear();
				this.#logger.info("Console cleared");
			}
		},
		{
			key: "o",
			description: "open in browser",
			handler: () => this.#handleOpenBrowser()
		},
		{
			key: "h",
			description: "show this help",
			handler: () => this.showHelp()
		}
	];
	constructor(options) {
		this.#logger = options.logger;
		this.#callbacks = options.callbacks;
	}
	setServerUrl(url) {
		this.#serverUrl = url;
	}
	setup() {
		if (!process.stdin.isTTY) return;
		process.stdin.setRawMode(true);
		this.#keyPressHandler = (data) => this.#handleKeyPress(data.toString());
		process.stdin.on("data", this.#keyPressHandler);
	}
	#handleKeyPress(key) {
		if (key === "" || key === "") return this.#callbacks.onQuit();
		const shortcut = this.#shortcuts.find((s) => s.key === key);
		if (shortcut) shortcut.handler();
	}
	async #handleOpenBrowser() {
		this.#logger.log("");
		this.#logger.info(`Opening ${this.#serverUrl}...`);
		const { default: open } = await import("open");
		open(this.#serverUrl);
	}
	showHelp() {
		this.#logger.log("");
		this.#logger.log("Available shortcuts:");
		this.#shortcuts.forEach(({ key, description }) => this.#logger.log(`· ${key}: ${description}`));
	}
	cleanup() {
		if (!process.stdin.isTTY) return;
		process.stdin.setRawMode(false);
		process.stdin.pause();
		process.stdin.unref();
		process.stdin.removeListener("data", this.#keyPressHandler);
		this.#keyPressHandler = void 0;
	}
};
var DevServer = class DevServer {
	static #HOT_HOOK_FULL_RELOAD_INFO = {
		source: "hot-hook",
		fullReload: true,
		hotReloaded: false
	};
	static #HOT_HOOK_INVALIDATED_INFO = {
		source: "hot-hook",
		fullReload: false,
		hotReloaded: true
	};
	static #WATCHER_INFO = {
		source: "watcher",
		fullReload: true,
		hotReloaded: false
	};
	#onError;
	#onClose;
	#stickyPort;
	#stickyHmrPort;
	#mode = "static";
	#watcher;
	#httpServer;
	#isHttpServerAlive = false;
	#shortcutsManager;
	#fileSystem;
	#indexGenerator;
	#routesScanner;
	#hooks;
	ui = cliui();
	#restartHTTPServer = throttle(async () => {
		if (this.#httpServer) {
			this.#httpServer.removeAllListeners();
			this.#httpServer.kill("SIGKILL");
		}
		await this.#startHTTPServer(this.#stickyPort, this.#stickyHmrPort);
	}, "restartHTTPServer");
	#setupKeyboardShortcuts() {
		this.#shortcutsManager = new ShortcutsManager({
			logger: this.ui.logger,
			callbacks: {
				onRestart: () => this.#restartHTTPServer(),
				onClear: () => this.#clearScreen(),
				onQuit: () => this.close()
			}
		});
		this.#shortcutsManager.setup();
	}
	#cleanupKeyboardShortcuts() {
		this.#shortcutsManager?.cleanup();
	}
	get mode() {
		return this.#mode;
	}
	scriptFile = "bin/server.ts";
	cwd;
	cwdPath;
	options;
	constructor(cwd, options) {
		this.cwd = cwd;
		this.options = options;
		this.cwdPath = string.toUnixSlash(fileURLToPath(this.cwd));
	}
	#isAdonisJSReadyMessage(message) {
		return message !== null && typeof message === "object" && "isAdonisJS" in message && "environment" in message && message.environment === "web";
	}
	#isAdonisJSRoutesMessage(message) {
		return message !== null && typeof message === "object" && "routesFileLocation" in message;
	}
	async #postServerReady(message) {
		const host = message.host === "0.0.0.0" ? "127.0.0.1" : message.host;
		const info = {
			host,
			port: message.port
		};
		const serverUrl = `http://${host}:${message.port}`;
		this.#shortcutsManager?.setServerUrl(serverUrl);
		const displayMessage = this.ui.sticker().add(`Server address: ${this.ui.colors.cyan(serverUrl)}`).add(`Mode: ${this.ui.colors.cyan(this.mode)}`);
		if (message.duration) displayMessage.add(`Ready in: ${this.ui.colors.cyan(prettyHrtime(message.duration))}`);
		displayMessage.add(`Press ${this.ui.colors.dim("h")} to show help`);
		try {
			await this.#hooks.runner("devServerStarted").run(this, info, displayMessage);
		} catch (error) {
			this.ui.logger.error("One of the \"devServerStarted\" hooks failed");
			this.ui.logger.fatal(error);
		}
		displayMessage.render();
	}
	#isHotHookMessage(message) {
		return message !== null && typeof message === "object" && "type" in message && typeof message.type === "string" && message.type.startsWith("hot-hook:");
	}
	#clearScreen() {
		if (this.options.clearScreen) process.stdout.write("\x1Bc");
	}
	#createWatcher(options) {
		const watcher = watch({
			usePolling: options?.poll ?? false,
			cwd: this.cwdPath,
			ignoreInitial: true,
			ignored: (file, stats) => {
				if (!stats) return false;
				if (file.includes("inertia") && !file.includes("node_modules")) return false;
				if (stats.isFile()) return !this.#fileSystem.shouldWatchFile(file);
				return !this.#fileSystem.shouldWatchDirectory(file);
			}
		});
		watcher.on("error", (error) => {
			this.ui.logger.warning("file system watcher failure");
			this.ui.logger.fatal(error);
			this.#onError?.(error);
			this.#watcher?.close();
		});
		watcher.on("ready", () => {
			this.ui.logger.info("watching file system for changes...");
		});
		return watcher;
	}
	#handleHmrWatcherEvent(options) {
		const relativePath = string.toUnixSlash(options.filePath);
		const absolutePath = join(this.cwdPath, relativePath);
		if (this.#isHttpServerAlive === false) {
			this.#clearScreen();
			this.ui.logger.log(`${this.ui.colors.green(options.displayLabel)} ${relativePath}`);
			this.#restartHTTPServer();
			return;
		}
		if (options.action === "add") this.#hooks.runner("fileAdded").run(relativePath, absolutePath, this);
		else if (options.action === "unlink") this.#hooks.runner("fileRemoved").run(relativePath, absolutePath, this);
		this.#httpServer?.send({
			type: "hot-hook:file-changed",
			path: absolutePath,
			action: options.action
		});
	}
	#handleFileChange(relativePath, absolutePath, action, info) {
		if ((action === "add" || action === "delete") && this.mode === "hmr") {
			debug_default("ignoring add and delete actions in HMR mode %s", relativePath);
			return;
		}
		if (info && info.source === "hot-hook" && info.hotReloaded) {
			debug_default("hot reloading %s, info %O", relativePath, info);
			this.ui.logger.log(`${this.ui.colors.green("invalidated")} ${relativePath}`);
			return;
		}
		if (info && !info.fullReload) {
			debug_default("ignoring full reload", relativePath, info);
			return;
		}
		const file = this.#fileSystem.inspect(absolutePath, relativePath);
		if (!file) return;
		if (file.reloadServer) {
			this.#clearScreen();
			this.ui.logger.log(`${this.ui.colors.green(action)} ${relativePath}`);
			this.#restartHTTPServer();
			return;
		}
		this.ui.logger.log(`${this.ui.colors.green(action)} ${relativePath}`);
	}
	#regenerateIndex(filePath, action) {
		if (action === "add") return this.#indexGenerator.addFile(filePath);
		return this.#indexGenerator.removeFile(filePath);
	}
	async #reScanRoutes(filePath) {
		if (!this.#routesScanner) return;
		try {
			if (await this.#routesScanner.invalidate(filePath)) await this.#hooks.runner("routesScanned").run(this, this.#routesScanner);
		} catch (error) {
			this.ui.logger.error("Unable to rescan routes because of the following error");
			this.ui.logger.fatal(error);
		}
	}
	#processRoutes = throttle(async (routesFileLocation) => {
		try {
			const scanRoutes = this.#hooks.has("routesScanning") || this.#hooks.has("routesScanned");
			const shareRoutes = this.#hooks.has("routesCommitted");
			if (!scanRoutes && !shareRoutes) {
				unlink(routesFileLocation).catch(() => {});
				return;
			}
			const routesJSON = await readFile(routesFileLocation, "utf-8");
			const routesList = JSON.parse(routesJSON);
			unlink(routesFileLocation).catch(() => {});
			if (shareRoutes) await this.#hooks.runner("routesCommitted").run(this, routesList);
			if (scanRoutes) {
				this.#routesScanner = new RoutesScanner(this.cwdPath, []);
				await this.#hooks.runner("routesScanning").run(this, this.#routesScanner);
				for (const domain of Object.keys(routesList)) await this.#routesScanner.scan(routesList[domain]);
				await this.#hooks.runner("routesScanned").run(this, this.#routesScanner);
			}
		} catch (error) {
			this.ui.logger.error("Unable to process and scan routes because of the following error");
			this.ui.logger.fatal(error);
		}
	}, "processRoutes");
	#registerServerRestartHooks() {
		this.#hooks.add("fileAdded", (relativePath, absolutePath) => {
			this.#regenerateIndex(absolutePath, "add");
			this.#handleFileChange(relativePath, absolutePath, "add");
		});
		this.#hooks.add("fileChanged", (relativePath, absolutePath, info) => {
			if (info.hotReloaded || !info.hotReloaded && !info.fullReload) this.#reScanRoutes(absolutePath);
			this.#handleFileChange(relativePath, absolutePath, "update", info);
		});
		this.#hooks.add("fileRemoved", (relativePath, absolutePath) => {
			this.#regenerateIndex(absolutePath, "delete");
			this.#handleFileChange(relativePath, absolutePath, "delete");
		});
	}
	async #init(mode) {
		const tsConfig = readTsConfig(this.cwdPath);
		if (!tsConfig) {
			this.#onError?.(new RuntimeException("Unable to parse tsconfig file"));
			return false;
		}
		this.#mode = mode;
		this.#clearScreen();
		this.ui.logger.info(`starting server in ${this.#mode} mode...`);
		this.#indexGenerator = new IndexGenerator(this.cwdPath, this.ui.logger);
		this.#stickyPort = String(await getPort(this.cwd));
		this.#stickyHmrPort = String(await getRandomPort({ port: 24678 }));
		this.#fileSystem = new FileSystem(this.cwdPath, tsConfig, this.options);
		this.ui.logger.info("loading hooks...");
		this.#hooks = await loadHooks(this.options.hooks, [
			"init",
			"routesCommitted",
			"routesScanning",
			"routesScanned",
			"devServerStarting",
			"devServerStarted",
			"fileAdded",
			"fileChanged",
			"fileRemoved"
		]);
		this.#registerServerRestartHooks();
		this.#setupKeyboardShortcuts();
		await this.#hooks.runner("init").run(this, this.#hooks, this.#indexGenerator);
		this.#hooks.clear("init");
		this.ui.logger.info("generating indexes...");
		await this.#indexGenerator.generate();
		return true;
	}
	async #startHTTPServer(port, hmrPort) {
		await this.#hooks.runner("devServerStarting").run(this);
		debug_default("starting http server using \"%s\" file, options %O", this.scriptFile, this.options);
		return new Promise((resolve) => {
			this.#httpServer = runNode(this.cwd, {
				script: this.scriptFile,
				env: {
					PORT: port,
					VITE_HMR_PORT: hmrPort,
					DEV_MODE: "true",
					...this.options.env
				},
				nodeArgs: this.options.nodeArgs,
				reject: true,
				scriptArgs: this.options.scriptArgs
			});
			this.#isHttpServerAlive = true;
			this.#httpServer.on("message", async (message) => {
				if (this.#isAdonisJSReadyMessage(message)) {
					debug_default("received http server ready message %O", message);
					await this.#postServerReady(message);
					resolve();
				} else if (this.#isAdonisJSRoutesMessage(message)) {
					debug_default("received routes location from the server %O", message);
					await this.#processRoutes(message.routesFileLocation);
				} else if (this.#mode === "hmr" && this.#isHotHookMessage(message)) {
					debug_default("received hot-hook message %O", message);
					if (message.type === "hot-hook:full-reload") {
						const absolutePath = message.path ? string.toUnixSlash(message.path) : "";
						const relativePath = relative(this.cwdPath, absolutePath);
						this.#hooks.runner("fileChanged").run(relativePath, absolutePath, DevServer.#HOT_HOOK_FULL_RELOAD_INFO, this);
					} else if (message.type === "hot-hook:invalidated") {
						const absolutePath = message.paths[0] ? string.toUnixSlash(message.paths[0]) : "";
						const relativePath = relative(this.cwdPath, absolutePath);
						this.#hooks.runner("fileChanged").run(relativePath, absolutePath, DevServer.#HOT_HOOK_INVALIDATED_INFO, this);
					}
				}
			});
			this.#httpServer.then((result) => {
				this.#isHttpServerAlive = false;
				if (!this.#watcher) this.#onClose?.(result.exitCode);
				else this.ui.logger.info("Underlying HTTP server closed. Still watching for changes");
			}).catch((error) => {
				this.#isHttpServerAlive = false;
				if (!this.#watcher) this.#onError?.(error);
				else this.ui.logger.info("Underlying HTTP server died. Still watching for changes");
			}).finally(() => {
				resolve();
			});
		});
	}
	onClose(callback) {
		this.#onClose = callback;
		return this;
	}
	onError(callback) {
		this.#onError = callback;
		return this;
	}
	async close() {
		this.#cleanupKeyboardShortcuts();
		await this.#watcher?.close();
		if (this.#httpServer) {
			this.#httpServer.removeAllListeners();
			this.#httpServer.kill("SIGKILL");
		}
	}
	async start() {
		if (!await this.#init(this.options.hmr ? "hmr" : "static")) return;
		if (this.#mode === "hmr") {
			this.options.nodeArgs.push("--import=hot-hook/register");
			this.options.env = {
				...this.options.env,
				HOT_HOOK_WATCH: "false"
			};
		}
		this.ui.logger.info("starting HTTP server...");
		await this.#startHTTPServer(this.#stickyPort, this.#stickyHmrPort);
		if (this.#mode !== "hmr") return;
		this.#watcher = this.#createWatcher();
		this.#watcher.on("add", (filePath) => {
			this.#handleHmrWatcherEvent({
				filePath,
				action: "add",
				displayLabel: "add"
			});
		});
		this.#watcher.on("change", (filePath) => {
			this.#handleHmrWatcherEvent({
				filePath,
				action: "change",
				displayLabel: "update"
			});
		});
		this.#watcher.on("unlink", (filePath) => {
			this.#handleHmrWatcherEvent({
				filePath,
				action: "unlink",
				displayLabel: "delete"
			});
		});
	}
	async startAndWatch(options) {
		if (!await this.#init("watch")) return;
		this.ui.logger.info("starting HTTP server...");
		await this.#startHTTPServer(this.#stickyPort, this.#stickyHmrPort);
		this.#watcher = this.#createWatcher({ poll: options?.poll });
		this.#watcher.on("add", (filePath) => {
			const relativePath = string.toUnixSlash(filePath);
			const absolutePath = join(this.cwdPath, relativePath);
			this.#hooks.runner("fileAdded").run(relativePath, absolutePath, this);
		});
		this.#watcher.on("change", (filePath) => {
			const relativePath = string.toUnixSlash(filePath);
			const absolutePath = join(this.cwdPath, relativePath);
			this.#hooks.runner("fileChanged").run(relativePath, absolutePath, DevServer.#WATCHER_INFO, this);
		});
		this.#watcher.on("unlink", (filePath) => {
			const relativePath = string.toUnixSlash(filePath);
			const absolutePath = join(this.cwdPath, relativePath);
			this.#hooks.runner("fileRemoved").run(relativePath, absolutePath, this);
		});
	}
};
var TestRunner = class {
	#onError;
	#onClose;
	#stickyPort;
	#stickyHmrPort;
	#watcher;
	#testsProcess;
	#fileSystem;
	#hooks;
	#indexGenerator;
	ui = cliui();
	#reRunTests = throttle(async (filters) => {
		if (this.#testsProcess) {
			this.#testsProcess.removeAllListeners();
			this.#testsProcess.kill("SIGKILL");
		}
		await this.#runTests(this.#stickyPort, this.#stickyHmrPort, filters);
	}, "reRunTests");
	scriptFile = "bin/test.ts";
	cwd;
	cwdPath;
	options;
	constructor(cwd, options) {
		this.cwd = cwd;
		this.options = options;
		this.cwdPath = string.toUnixSlash(fileURLToPath(this.cwd));
	}
	#convertOptionsToArgs() {
		const args = [];
		if (this.options.reporters) {
			args.push("--reporters");
			args.push(this.options.reporters.join(","));
		}
		if (this.options.timeout !== void 0) {
			args.push("--timeout");
			args.push(String(this.options.timeout));
		}
		if (this.options.failed) args.push("--failed");
		if (this.options.retries !== void 0) {
			args.push("--retries");
			args.push(String(this.options.retries));
		}
		return args;
	}
	#convertFiltersToArgs(filters) {
		const args = [];
		if (filters.suites) args.push(...filters.suites);
		if (filters.files) {
			args.push("--files");
			args.push(filters.files.join(","));
		}
		if (filters.groups) {
			args.push("--groups");
			args.push(filters.groups.join(","));
		}
		if (filters.tags) {
			args.push("--tags");
			args.push(filters.tags.join(","));
		}
		if (filters.tests) {
			args.push("--tests");
			args.push(filters.tests.join(","));
		}
		return args;
	}
	#clearScreen() {
		if (this.options.clearScreen) process.stdout.write("\x1Bc");
	}
	async #runTests(port, hmrPort, filters) {
		await this.#hooks.runner("testsStarting").run(this);
		debug_default("running tests using \"%s\" file, options %O", this.scriptFile, this.options);
		return new Promise(async (resolve) => {
			const mergedFilters = {
				...this.options.filters,
				...filters
			};
			const scriptArgs = [
				...this.#convertOptionsToArgs(),
				...this.options.scriptArgs,
				...this.#convertFiltersToArgs(mergedFilters)
			];
			this.#testsProcess = runNode(this.cwd, {
				script: this.scriptFile,
				reject: true,
				env: {
					PORT: port,
					VITE_HMR_PORT: hmrPort,
					...this.options.env
				},
				nodeArgs: this.options.nodeArgs,
				scriptArgs
			});
			this.#testsProcess.then((result) => {
				this.#hooks.runner("testsFinished").run(this).catch((error) => {
					this.ui.logger.error("One of the \"testsFinished\" hooks failed");
					this.ui.logger.fatal(error);
				}).finally(() => {
					if (!this.#watcher) {
						this.#onClose?.(result.exitCode);
						this.close();
					}
				});
			}).catch((error) => {
				if (!this.#watcher) {
					this.#onError?.(error);
					this.close();
				} else this.ui.logger.info("Underlying HTTP server died. Still watching for changes");
			}).finally(() => resolve());
		});
	}
	#handleFileChange(relativePath, absolutePath, action) {
		const file = this.#fileSystem.inspect(absolutePath, relativePath);
		if (!file) return;
		this.#clearScreen();
		this.ui.logger.log(`${this.ui.colors.green(action)} ${relativePath}`);
		if (file.fileType === "test") this.#reRunTests({ files: [relativePath] });
		else this.#reRunTests();
	}
	#regenerateIndex(filePath, action) {
		if (action === "add") return this.#indexGenerator.addFile(filePath);
		return this.#indexGenerator.removeFile(filePath);
	}
	#registerServerRestartHooks() {
		this.#hooks.add("fileAdded", (relativePath, absolutePath) => {
			this.#regenerateIndex(absolutePath, "add");
			this.#handleFileChange(relativePath, absolutePath, "add");
		});
		this.#hooks.add("fileChanged", (relativePath, absolutePath) => {
			this.#regenerateIndex(absolutePath, "add");
			this.#handleFileChange(relativePath, absolutePath, "update");
		});
		this.#hooks.add("fileRemoved", (relativePath, absolutePath) => {
			this.#regenerateIndex(absolutePath, "delete");
			this.#handleFileChange(relativePath, absolutePath, "delete");
		});
	}
	onClose(callback) {
		this.#onClose = callback;
		return this;
	}
	onError(callback) {
		this.#onError = callback;
		return this;
	}
	async close() {
		await this.#watcher?.close();
		if (this.#testsProcess) {
			this.#testsProcess.removeAllListeners();
			this.#testsProcess.kill("SIGKILL");
		}
	}
	async run() {
		this.#stickyPort = String(await getPort(this.cwd));
		this.#stickyHmrPort = String(await getRandomPort({ port: 24678 }));
		this.#indexGenerator = new IndexGenerator(this.cwdPath, this.ui.logger);
		this.#clearScreen();
		this.ui.logger.info("loading hooks...");
		this.#hooks = await loadHooks(this.options.hooks, [
			"init",
			"testsStarting",
			"testsFinished"
		]);
		await this.#hooks.runner("init").run(this, this.#hooks, this.#indexGenerator);
		this.#hooks.clear("init");
		this.ui.logger.info("generating indexes...");
		await this.#indexGenerator.generate();
		this.ui.logger.info("booting application to run tests...");
		await this.#runTests(this.#stickyPort, this.#stickyHmrPort);
	}
	async runAndWatch(options) {
		const tsConfig = readTsConfig(this.cwdPath);
		if (!tsConfig) {
			this.#onError?.(new RuntimeException("Unable to parse tsconfig file"));
			return;
		}
		this.#stickyPort = String(await getPort(this.cwd));
		this.#stickyHmrPort = String(getRandomPort({ port: 24678 }));
		this.#indexGenerator = new IndexGenerator(this.cwdPath, this.ui.logger);
		this.#fileSystem = new FileSystem(this.cwdPath, tsConfig, {
			...this.options,
			suites: this.options.suites?.filter((suite) => {
				if (this.options.filters.suites) return this.options.filters.suites.includes(suite.name);
				return true;
			})
		});
		this.#clearScreen();
		this.ui.logger.info("loading hooks...");
		this.#hooks = await loadHooks(this.options.hooks, [
			"init",
			"testsStarting",
			"testsFinished",
			"fileAdded",
			"fileChanged",
			"fileRemoved"
		]);
		this.#registerServerRestartHooks();
		await this.#hooks.runner("init").run(this, this.#hooks, this.#indexGenerator);
		this.#hooks.clear("init");
		this.ui.logger.info("generating indexes...");
		await this.#indexGenerator.generate();
		this.ui.logger.info("booting application to run tests...");
		await this.#runTests(this.#stickyPort, this.#stickyHmrPort);
		this.#watcher = watch({
			usePolling: options?.poll ?? false,
			cwd: this.cwdPath,
			ignoreInitial: true,
			ignored: (file, stats) => {
				if (!stats) return false;
				if (stats.isFile()) return !this.#fileSystem.shouldWatchFile(file);
				return !this.#fileSystem.shouldWatchDirectory(file);
			}
		});
		this.#watcher.on("ready", () => {
			this.ui.logger.info("watching file system for changes...");
		});
		this.#watcher.on("error", (error) => {
			this.ui.logger.warning("file system watcher failure");
			this.ui.logger.fatal(error);
			this.#onError?.(error);
			this.#watcher?.close();
		});
		this.#watcher.on("add", (filePath) => {
			const relativePath = string.toUnixSlash(filePath);
			const absolutePath = join(this.cwdPath, filePath);
			this.#hooks.runner("fileAdded").run(relativePath, absolutePath, this);
		});
		this.#watcher.on("change", (filePath) => {
			const relativePath = string.toUnixSlash(filePath);
			const absolutePath = join(this.cwdPath, filePath);
			this.#hooks.runner("fileChanged").run(relativePath, absolutePath, {
				source: "watcher",
				fullReload: true,
				hotReloaded: false
			}, this);
		});
		this.#watcher.on("unlink", (filePath) => {
			const relativePath = string.toUnixSlash(filePath);
			const absolutePath = join(this.cwdPath, filePath);
			this.#hooks.runner("fileRemoved").run(relativePath, absolutePath, this);
		});
	}
};
export { Bundler, CodemodException, DevServer, FileBuffer, SUPPORTED_PACKAGE_MANAGERS, TestRunner, VirtualFileSystem };
