// @adonisjs/assembler/build/source-dVeugJ0e.js
import { f as throttle, l as removeExtension, m as debug_default, t as VirtualFileSystem } from "./virtual_file_system-bGeoWsK-.js";
import { mkdir, writeFile } from "node:fs/promises";
import string from "@poppinss/utils/string";
import { dirname, join, relative } from "node:path/posix";
import StringBuilder from "@poppinss/utils/string_builder";
var FileBuffer = class FileBuffer {
	#eol = false;
	#buffer = [];
	#identationSize = 0;
	#compiledOutput;
	create() {
		return new FileBuffer();
	}
	get size() {
		return this.#buffer.length;
	}
	eol(enabled) {
		this.#eol = enabled;
		return this;
	}
	writeLine(text) {
		if (typeof text === "string") this.#buffer.push(`${" ".repeat(this.#identationSize)}${text}\n`);
		else {
			this.#buffer.push(text);
			this.#buffer.push("");
		}
		return this;
	}
	write(text) {
		if (typeof text === "string") this.#buffer.push(`${" ".repeat(this.#identationSize)}${text}`);
		else this.#buffer.push(text);
		return this;
	}
	indent() {
		this.#identationSize += 2;
		return this;
	}
	dedent() {
		this.#identationSize -= 2;
		if (this.#identationSize < 0) this.#identationSize = 0;
		return this;
	}
	toString() {
		return this.flush();
	}
	flush() {
		if (this.#compiledOutput !== void 0) return this.#compiledOutput;
		this.#compiledOutput = this.#buffer.join("\n");
		return this.#eol ? `${this.#compiledOutput}\n` : this.#compiledOutput;
	}
};
var IndexGeneratorSource = class {
	#appRoot;
	#output;
	#source;
	#outputDirname;
	#vfs;
	#config;
	#cliLogger;
	#generateOutput = throttle(async () => {
		const buffer = new FileBuffer().eol(true);
		if (this.#config.comment) buffer.writeLine(this.#textToComment(typeof this.#config.comment === "string" ? this.#config.comment : `This file is automatically generated.\nDO NOT EDIT manually`));
		if (this.#config.as === "barrelFile") this.#asBarrelFile(this.#vfs, buffer, this.#config.exportName, this.#config.disableLazyImports);
		else this.#config.as(this.#vfs, buffer, this.#config, { toImportPath: this.#createBarrelFileImportGenerator(this.#source, this.#outputDirname, this.#config) });
		await mkdir(dirname(this.#output), { recursive: true });
		await writeFile(this.#output, buffer.flush());
	});
	name;
	constructor(name, appRoot, cliLogger, config) {
		this.name = name;
		this.#config = config;
		this.#appRoot = appRoot;
		this.#cliLogger = cliLogger;
		this.#source = join(this.#appRoot, this.#config.source);
		this.#output = join(this.#appRoot, this.#config.output);
		this.#outputDirname = dirname(this.#output);
		this.#vfs = new VirtualFileSystem(this.#source, {
			glob: this.#config.glob,
			filter: this.#config.filter
		});
	}
	#treeToString(input, buffer) {
		Object.keys(input).forEach((key) => {
			const value = input[key];
			if (typeof value === "string") buffer.write(`${key}: ${value},`);
			else {
				buffer.write(`${key}: {`).indent();
				this.#treeToString(value, buffer);
				buffer.dedent().write(`},`);
			}
		});
	}
	#textToComment(text) {
		return `/**\n * ${text.split("\n").join("\n * ")}\n */`;
	}
	#createBarrelFileKeyGenerator(config) {
		return function(key) {
			let paths = key.split("/");
			let baseName = new StringBuilder(paths.pop());
			if (config.skipSegments?.length) paths = paths.filter((p) => !config.skipSegments.includes(p));
			if (config.computeBaseName) baseName = config.computeBaseName(baseName);
			else baseName = baseName.removeSuffix(config.removeSuffix ?? "").pascalCase();
			return [...paths.map((p) => string.camelCase(p)), baseName.toString()].join("/");
		};
	}
	#createBarrelFileImportGenerator(source, outputDirname, config) {
		return function(filePath) {
			if (config.importAlias) {
				debug_default("converting \"%s\" to import alias, source \"%s\"", filePath, source);
				return removeExtension(filePath.replace(source, config.importAlias));
			}
			debug_default("converting \"%s\" to relative import, source \"%s\"", filePath, outputDirname);
			return relative(outputDirname, filePath);
		};
	}
	#asBarrelFile(vfs, buffer, exportName, disableLazyImports) {
		const useEagerImports = disableLazyImports === true ? true : false;
		const keyGenerator = this.#createBarrelFileKeyGenerator(this.#config);
		const importsBuffer = buffer.create();
		const importGenerator = this.#createBarrelFileImportGenerator(this.#source, this.#outputDirname, this.#config);
		const tree = vfs.asTree({
			transformKey: keyGenerator,
			transformValue: (filePath, key) => {
				if (useEagerImports) {
					const importKey = new StringBuilder(key).pascalCase().toString();
					importsBuffer.write(`import ${importKey} from '${importGenerator(filePath)}'`);
					return importKey;
				}
				return `() => import('${importGenerator(filePath)}')`;
			}
		});
		if (!Object.keys(tree).length) {
			buffer.write(`export const ${exportName} = {}`);
			return;
		}
		if (useEagerImports) buffer.writeLine(importsBuffer);
		buffer.write(`export const ${exportName} = {`).indent();
		this.#treeToString(tree, buffer);
		buffer.dedent().write(`}`);
	}
	#logCreation(startTime) {
		this.#cliLogger.info(`created ${this.#config.output}`, { startTime });
	}
	async addFile(filePath) {
		if (this.#vfs.add(filePath)) {
			debug_default("file added, re-generating \"%s\" index", this.name);
			const startTime = process.hrtime();
			await this.#generateOutput();
			this.#logCreation(startTime);
		}
	}
	async removeFile(filePath) {
		if (this.#vfs.remove(filePath)) {
			debug_default("file removed, re-generating \"%s\" index", this.name);
			const startTime = process.hrtime();
			await this.#generateOutput();
			this.#logCreation(startTime);
		}
	}
	async generate() {
		debug_default("generating \"%s\" index", this.name);
		await this.#vfs.scan();
		await this.#generateOutput();
	}
};
export { FileBuffer as n, IndexGeneratorSource as t };
