// @adonisjs/assembler/build/validator_extractor-Ccio_Ndi.js
import { i as isRelative, m as debug_default } from "./virtual_file_system-bGeoWsK-.js";
import { findImport, inspectClass, inspectClassMethods, inspectMethodArguments, nodeToPlainText, searchValidatorDirectUsage } from "./src/helpers.js";
import { fileURLToPath, pathToFileURL } from "node:url";
import string from "@poppinss/utils/string";
import { join, relative } from "node:path";
import { resolve } from "import-meta-resolve";
var PathsResolver = class {
	#appRoot;
	#resolvedPaths = {};
	#resolver = (specifier, parentPath) => resolve(specifier, parentPath);
	constructor(appRoot) {
		this.#appRoot = pathToFileURL(join(appRoot, "index.js")).href;
	}
	use(resolver) {
		this.#resolver = resolver;
	}
	resolve(specifier, rewriteAliasImportExtension = false) {
		if (isRelative(specifier)) throw new Error("Cannot resolve relative paths using PathsResolver");
		const cacheKey = specifier;
		const cached = this.#resolvedPaths[cacheKey];
		if (cached) return cached;
		let resolvedPath = fileURLToPath(this.#resolver(specifier, this.#appRoot));
		if (rewriteAliasImportExtension && specifier.startsWith("#") && resolvedPath.endsWith(".js")) resolvedPath = resolvedPath.replace(/\.js$/, ".ts");
		this.#resolvedPaths[cacheKey] = resolvedPath;
		return this.#resolvedPaths[cacheKey];
	}
};
async function extractValidators(appRoot, vfs, controller) {
	const root = await vfs.get(controller.path);
	const fileContents = root.text();
	const controllerClass = inspectClass(root);
	if (!controllerClass) {
		debug_default(`No class defined within the "%s"`, controller.import.specifier);
		return;
	}
	const method = inspectClassMethods(controllerClass).find((methodNode) => {
		return methodNode.find({ rule: { kind: "property_identifier" } })?.text() === controller.method;
	});
	if (!method) {
		debug_default(`Unable to find "%s" method in "%s"`, controller.method, controller.import.specifier);
		return;
	}
	const validationCalls = inspectMethodArguments(method, [
		"request.validateUsing",
		"$CTX.request.validateUsing",
		"vine.validate"
	]).map((node) => {
		const firstArg = node.find({ rule: { any: [{ kind: "identifier" }, { kind: "member_expression" }] } });
		if (!firstArg) return;
		return nodeToPlainText(firstArg);
	}).filter((node) => node !== void 0).concat(searchValidatorDirectUsage(method).map((node) => nodeToPlainText(node)));
	if (!validationCalls.length) {
		debug_default(`Unable to detect any validation calls in "%s.%s" method`, controller.import.specifier, controller.method);
		return;
	}
	const controllerName = controllerClass.find({ rule: { kind: "type_identifier" } }).text();
	const controllerURL = pathToFileURL(controller.path);
	return (await Promise.all(validationCalls.map(async (validationCall) => {
		if (validationCall.split(".")[0] === controllerName) return {
			name: validationCall,
			import: {
				specifier: controller.import.specifier,
				type: "default",
				value: controllerName
			}
		};
		const importCall = await findImport(fileContents, validationCall);
		if (!importCall) {
			debug_default("Unable to find import for \"%s\" used by \"%s.%s\" method", validationCall, controller.import.specifier, controller.method);
			return null;
		}
		return {
			name: validationCall,
			import: {
				specifier: isRelative(importCall.specifier) ? string.toUnixSlash(relative(appRoot, fileURLToPath(new URL(importCall.specifier, controllerURL)))) : importCall.specifier,
				type: importCall.clause.type,
				value: importCall.clause.value
			}
		};
	}))).filter((value) => !!value);
}
export { PathsResolver as n, extractValidators as t };
